// symbol-openapi-typescript-fetch-client/dist/apis/MosaicRoutesApi.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MosaicRoutesApi = void 0;
const runtime = require("../runtime");
const models_1 = require("../models");
/**
 *
 */
class MosaicRoutesApi extends runtime.BaseAPI {
    /**
     * Gets the mosaic definition for a given mosaic identifier.
     * Get mosaic information
     */
    getMosaicRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            if (requestParameters.mosaicId === null || requestParameters.mosaicId === undefined) {
                throw new runtime.RequiredError('mosaicId', 'Required parameter requestParameters.mosaicId was null or undefined when calling getMosaic.');
            }
            const queryParameters = {};
            const headerParameters = {};
            const response = yield this.request({
                path: `/mosaics/{mosaicId}`.replace(`{${"mosaicId"}}`, encodeURIComponent(String(requestParameters.mosaicId))),
                method: 'GET',
                headers: headerParameters,
                query: queryParameters,
            });
            return new runtime.JSONApiResponse(response, (jsonValue) => models_1.MosaicInfoDTOFromJSON(jsonValue));
        });
    }
    /**
     * Gets the mosaic definition for a given mosaic identifier.
     * Get mosaic information
     */
    getMosaic(mosaicId) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.getMosaicRaw({ mosaicId: mosaicId });
            return yield response.value();
        });
    }
    /**
     * Gets an array of mosaic definition.
     * Get mosaics information for an array of mosaics
     */
    getMosaicsRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            if (requestParameters.mosaicIds === null || requestParameters.mosaicIds === undefined) {
                throw new runtime.RequiredError('mosaicIds', 'Required parameter requestParameters.mosaicIds was null or undefined when calling getMosaics.');
            }
            const queryParameters = {};
            const headerParameters = {};
            headerParameters['Content-Type'] = 'application/json';
            const response = yield this.request({
                path: `/mosaics`,
                method: 'POST',
                headers: headerParameters,
                query: queryParameters,
                body: models_1.MosaicIdsToJSON(requestParameters.mosaicIds),
            });
            return new runtime.JSONApiResponse(response, (jsonValue) => jsonValue.map(models_1.MosaicInfoDTOFromJSON));
        });
    }
    /**
     * Gets an array of mosaic definition.
     * Get mosaics information for an array of mosaics
     */
    getMosaics(mosaicIds) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.getMosaicsRaw({ mosaicIds: mosaicIds });
            return yield response.value();
        });
    }
    /**
     * Gets an array of mosaics.
     * Search mosaics
     */
    searchMosaicsRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            const queryParameters = {};
            if (requestParameters.ownerAddress !== undefined) {
                queryParameters['ownerAddress'] = requestParameters.ownerAddress;
            }
            if (requestParameters.pageSize !== undefined) {
                queryParameters['pageSize'] = requestParameters.pageSize;
            }
            if (requestParameters.pageNumber !== undefined) {
                queryParameters['pageNumber'] = requestParameters.pageNumber;
            }
            if (requestParameters.offset !== undefined) {
                queryParameters['offset'] = requestParameters.offset;
            }
            if (requestParameters.order !== undefined) {
                queryParameters['order'] = requestParameters.order;
            }
            const headerParameters = {};
            const response = yield this.request({
                path: `/mosaics`,
                method: 'GET',
                headers: headerParameters,
                query: queryParameters,
            });
            return new runtime.JSONApiResponse(response, (jsonValue) => models_1.MosaicPageFromJSON(jsonValue));
        });
    }
    /**
     * Gets an array of mosaics.
     * Search mosaics
     */
    searchMosaics(ownerAddress, pageSize, pageNumber, offset, order) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.searchMosaicsRaw({ ownerAddress: ownerAddress, pageSize: pageSize, pageNumber: pageNumber, offset: offset, order: order });
            return yield response.value();
        });
    }
}
exports.MosaicRoutesApi = MosaicRoutesApi;
