// symbol-sdk/dist/src/model/network/MosaicNetworkProperties.js
"use strict";
/*
 * Copyright 2020 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
class MosaicNetworkProperties {
    /**
     * @param maxMosaicsPerAccount - Maximum number of mosaics that an account can own.
     * @param maxMosaicDuration - Maximum mosaic duration.
     * @param maxMosaicDivisibility - Maximum mosaic divisibility.
     * @param mosaicRentalFeeSinkAddress - Public key of the mosaic rental fee sink address.
     * @param mosaicRentalFee - Mosaic rental fee.
     */
    constructor(maxMosaicsPerAccount, maxMosaicDuration, maxMosaicDivisibility, mosaicRentalFeeSinkAddress, mosaicRentalFee) {
        this.maxMosaicsPerAccount = maxMosaicsPerAccount;
        this.maxMosaicDuration = maxMosaicDuration;
        this.maxMosaicDivisibility = maxMosaicDivisibility;
        this.mosaicRentalFeeSinkAddress = mosaicRentalFeeSinkAddress;
        this.mosaicRentalFee = mosaicRentalFee;
    }
}
exports.MosaicNetworkProperties = MosaicNetworkProperties;
