// symbol-sdk/dist/src/infrastructure/SecretLockHttp.js
"use strict";
/*
 * Copyright 2020 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const symbol_openapi_typescript_fetch_client_1 = require("symbol-openapi-typescript-fetch-client");
const Address_1 = require("../model/account/Address");
const MosaicId_1 = require("../model/mosaic/MosaicId");
const UInt64_1 = require("../model/UInt64");
const Http_1 = require("./Http");
const DtoMapping_1 = require("../core/utils/DtoMapping");
const SecretLockInfo_1 = require("../model/lock/SecretLockInfo");
/**
 * SecretLock http repository.
 *
 * @since 1.0
 */
class SecretLockHttp extends Http_1.Http {
    /**
     * Constructor
     * @param url Base catapult-rest url
     * @param fetchApi fetch function to be used when performing rest requests.
     */
    constructor(url, fetchApi) {
        super(url, fetchApi);
        this.secretLockRoutesApi = new symbol_openapi_typescript_fetch_client_1.SecretLockRoutesApi(this.config());
    }
    /**
     * Get secret lock info for an account.
     * @param secret SecretLock secret
     * @returns Observable<SecretLockInfo>
     */
    getSecretLock(secret) {
        return this.call(this.secretLockRoutesApi.getSecretLock(secret), (body) => this.toSecretLockInfo(body));
    }
    /**
     * Gets an array of SecretLockInfo.
     * @param criteria - SecretLock search criteria
     * @returns Observable<Page<SecretLockInfo>>
     */
    search(criteria) {
        return this.call(this.secretLockRoutesApi.searchSecretLock(criteria.address.plain(), criteria.pageSize, criteria.pageNumber, criteria.offset, DtoMapping_1.DtoMapping.mapEnum(criteria.order)), (body) => super.toPage(body.pagination, body.data, this.toSecretLockInfo));
    }
    /**
     * This method maps a SecretLockInfoDTO from rest to the SDK's SecretLockInfo model object.
     *
     * @internal
     * @param {SecretLockInfoDTO} dto SecretLockInfoDTO the dto object from rest.
     * @returns SecretLockInfo model
     */
    toSecretLockInfo(dto) {
        return new SecretLockInfo_1.SecretLockInfo(dto.id, Address_1.Address.createFromEncoded(dto.lock.ownerAddress), new MosaicId_1.MosaicId(dto.lock.mosaicId), UInt64_1.UInt64.fromNumericString(dto.lock.amount), UInt64_1.UInt64.fromNumericString(dto.lock.endHeight), dto.lock.status, dto.lock.hashAlgorithm.valueOf(), dto.lock.secret, Address_1.Address.createFromEncoded(dto.lock.recipientAddress), dto.lock.compositeHash);
    }
}
exports.SecretLockHttp = SecretLockHttp;
