// @geens_npo/defined-id/lib/handlers/did/GenerateDidDocumentHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.GenerateDidDocumentHandler = void 0;
const __1 = require("../..");
const BaseHandler_1 = require("../BaseHandler");
class GenerateDidDocumentHandler extends BaseHandler_1.BaseHandler {
    handleCore(request) {
        const didDocument = {
            context: ['https://w3id.org/did/v1', 'https://w3id.org/security/v1'],
            id: request.did,
            salt: request.salt,
            publicKey: request.pubKey,
            service: request.services,
            created: (new Date()).toISOString()
        };
        return {
            didDocument: didDocument
        };
    }
    validateCore(request) {
        const errors = [];
        if (!request.did)
            errors.push('Did is required');
        else if (!__1.DidValidator.isDidStringValid(request.did))
            errors.push('Did is invalid');
        if (!request.salt)
            errors.push('Salt is required');
        else if (!__1.DidValidator.isSaltValid(request.salt))
            errors.push('Salt is invalid');
        if (!request.pubKey || request.pubKey.length === 0)
            errors.push('PubKey is required');
        if (!request.services || request.services.length === 0)
            errors.push('Services are required');
        return errors;
    }
}
exports.GenerateDidDocumentHandler = GenerateDidDocumentHandler;
