// symbol-sdk/dist/src/model/receipt/BalanceTransferReceipt.js
"use strict";
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const catbuffer_typescript_1 = require("catbuffer-typescript");
const Convert_1 = require("../../core/format/Convert");
const Receipt_1 = require("./Receipt");
const ReceiptVersion_1 = require("./ReceiptVersion");
/**
 * Balance Transfer: A mosaic transfer was triggered.
 */
class BalanceTransferReceipt extends Receipt_1.Receipt {
    /**
     * Balance transfer expiry receipt
     * @param senderAddress - The sender address.
     * @param recipientAddress - The mosaic recipient address.
     * @param mosaicId - The mosaic id.
     * @param amount - The amount of mosaic.
     * @param version - The receipt version
     * @param type - The receipt type
     * @param size - the receipt size
     */
    constructor(
    /**
     * The sender address.
     */
    senderAddress, 
    /**
     * The mosaic recipient address.
     */
    recipientAddress, 
    /**
     * The mosaic id.
     */
    mosaicId, 
    /**
     * The amount of mosaic.
     */
    amount, version, type, size) {
        super(version, type, size);
        this.senderAddress = senderAddress;
        this.recipientAddress = recipientAddress;
        this.mosaicId = mosaicId;
        this.amount = amount;
    }
    /**
     * @internal
     * Generate buffer
     * @return {Uint8Array}
     */
    serialize() {
        return new catbuffer_typescript_1.BalanceTransferReceiptBuilder(ReceiptVersion_1.ReceiptVersion.BALANCE_TRANSFER, this.type.valueOf(), new catbuffer_typescript_1.MosaicBuilder(new catbuffer_typescript_1.MosaicIdDto(this.mosaicId.toDTO()), new catbuffer_typescript_1.AmountDto(this.amount.toDTO())), new catbuffer_typescript_1.AddressDto(Convert_1.Convert.hexToUint8(this.senderAddress.encoded())), new catbuffer_typescript_1.AddressDto(Convert_1.Convert.hexToUint8(this.recipientAddress.encoded()))).serialize();
    }
}
exports.BalanceTransferReceipt = BalanceTransferReceipt;
