// symbol-sdk/dist/src/model/network/PluginProperties.js
"use strict";
/*
import { AccountLinkNetworkProperties } from './AccountLinkNetworkProperties';
 * Copyright 2020 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * Network related configuration properties.
 */
class PluginProperties {
    /**
     * @param accountlink - Network identifier.
     * @param aggregate - Nemesis public key.
     * @param lockhash - Nemesis generation hash.
     * @param locksecret - Nemesis epoch time adjustment.
     * @param metadata -
     * @param mosaic -
     * @param multisig -
     * @param namespace -
     * @param restrictionaccount -
     * @param restrictionmosaic -
     * @param transfer -
     */
    constructor(accountlink, aggregate, lockhash, locksecret, metadata, mosaic, multisig, namespace, restrictionaccount, restrictionmosaic, transfer) {
        this.accountlink = accountlink;
        this.aggregate = aggregate;
        this.lockhash = lockhash;
        this.locksecret = locksecret;
        this.metadata = metadata;
        this.mosaic = mosaic;
        this.multisig = multisig;
        this.namespace = namespace;
        this.restrictionaccount = restrictionaccount;
        this.restrictionmosaic = restrictionmosaic;
        this.transfer = transfer;
    }
}
exports.PluginProperties = PluginProperties;
