// symbol-openapi-typescript-fetch-client/dist/models/AccountTypeEnum.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountTypeEnumToJSON = exports.AccountTypeEnumFromJSONTyped = exports.AccountTypeEnumFromJSON = exports.AccountTypeEnum = void 0;
/**
 * * 0 - Unlinked.
 * * 1 - Balance-holding account that is linked to a remote harvester account.
 * * 2 - Remote harvester account that is linked to a balance-holding account.
 * * 3 - Remote harvester eligible account that is unlinked.
 * @export
 * @enum {string}
 */
var AccountTypeEnum;
(function (AccountTypeEnum) {
    AccountTypeEnum[AccountTypeEnum["NUMBER_0"] = 0] = "NUMBER_0";
    AccountTypeEnum[AccountTypeEnum["NUMBER_1"] = 1] = "NUMBER_1";
    AccountTypeEnum[AccountTypeEnum["NUMBER_2"] = 2] = "NUMBER_2";
    AccountTypeEnum[AccountTypeEnum["NUMBER_3"] = 3] = "NUMBER_3";
})(AccountTypeEnum = exports.AccountTypeEnum || (exports.AccountTypeEnum = {}));
function AccountTypeEnumFromJSON(json) {
    return AccountTypeEnumFromJSONTyped(json, false);
}
exports.AccountTypeEnumFromJSON = AccountTypeEnumFromJSON;
function AccountTypeEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.AccountTypeEnumFromJSONTyped = AccountTypeEnumFromJSONTyped;
function AccountTypeEnumToJSON(value) {
    return value;
}
exports.AccountTypeEnumToJSON = AccountTypeEnumToJSON;
