// symbol-openapi-typescript-fetch-client/dist/models/MessageTypeEnum.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MessageTypeEnumToJSON = exports.MessageTypeEnumFromJSONTyped = exports.MessageTypeEnumFromJSON = exports.MessageTypeEnum = void 0;
/**
 * Type of message:
 * * 0 - Regular message.
 * * 1 - Encrypted message.
 * * 254 - Persistent harvesting delegation.
 * @export
 * @enum {string}
 */
var MessageTypeEnum;
(function (MessageTypeEnum) {
    MessageTypeEnum[MessageTypeEnum["NUMBER_0"] = 0] = "NUMBER_0";
    MessageTypeEnum[MessageTypeEnum["NUMBER_1"] = 1] = "NUMBER_1";
    MessageTypeEnum[MessageTypeEnum["NUMBER_254"] = 254] = "NUMBER_254";
})(MessageTypeEnum = exports.MessageTypeEnum || (exports.MessageTypeEnum = {}));
function MessageTypeEnumFromJSON(json) {
    return MessageTypeEnumFromJSONTyped(json, false);
}
exports.MessageTypeEnumFromJSON = MessageTypeEnumFromJSON;
function MessageTypeEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.MessageTypeEnumFromJSONTyped = MessageTypeEnumFromJSONTyped;
function MessageTypeEnumToJSON(value) {
    return value;
}
exports.MessageTypeEnumToJSON = MessageTypeEnumToJSON;
