// symbol-openapi-typescript-fetch-client/dist/models/MosaicSupplyChangeActionEnum.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MosaicSupplyChangeActionEnumToJSON = exports.MosaicSupplyChangeActionEnumFromJSONTyped = exports.MosaicSupplyChangeActionEnumFromJSON = exports.MosaicSupplyChangeActionEnum = void 0;
/**
 * Direction of the supply change:
 * * 0  - Decrease.
 * * 1  - Increase.
 * @export
 * @enum {string}
 */
var MosaicSupplyChangeActionEnum;
(function (MosaicSupplyChangeActionEnum) {
    MosaicSupplyChangeActionEnum[MosaicSupplyChangeActionEnum["NUMBER_0"] = 0] = "NUMBER_0";
    MosaicSupplyChangeActionEnum[MosaicSupplyChangeActionEnum["NUMBER_1"] = 1] = "NUMBER_1";
})(MosaicSupplyChangeActionEnum = exports.MosaicSupplyChangeActionEnum || (exports.MosaicSupplyChangeActionEnum = {}));
function MosaicSupplyChangeActionEnumFromJSON(json) {
    return MosaicSupplyChangeActionEnumFromJSONTyped(json, false);
}
exports.MosaicSupplyChangeActionEnumFromJSON = MosaicSupplyChangeActionEnumFromJSON;
function MosaicSupplyChangeActionEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.MosaicSupplyChangeActionEnumFromJSONTyped = MosaicSupplyChangeActionEnumFromJSONTyped;
function MosaicSupplyChangeActionEnumToJSON(value) {
    return value;
}
exports.MosaicSupplyChangeActionEnumToJSON = MosaicSupplyChangeActionEnumToJSON;
