// symbol-openapi-typescript-fetch-client/dist/models/NamespaceRegistrationTypeEnum.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.NamespaceRegistrationTypeEnumToJSON = exports.NamespaceRegistrationTypeEnumFromJSONTyped = exports.NamespaceRegistrationTypeEnumFromJSON = exports.NamespaceRegistrationTypeEnum = void 0;
/**
 * Type of namespace:
 * * 0 - Root namespace.
 * * 1 - Subnamespace.
 * @export
 * @enum {string}
 */
var NamespaceRegistrationTypeEnum;
(function (NamespaceRegistrationTypeEnum) {
    NamespaceRegistrationTypeEnum[NamespaceRegistrationTypeEnum["NUMBER_0"] = 0] = "NUMBER_0";
    NamespaceRegistrationTypeEnum[NamespaceRegistrationTypeEnum["NUMBER_1"] = 1] = "NUMBER_1";
})(NamespaceRegistrationTypeEnum = exports.NamespaceRegistrationTypeEnum || (exports.NamespaceRegistrationTypeEnum = {}));
function NamespaceRegistrationTypeEnumFromJSON(json) {
    return NamespaceRegistrationTypeEnumFromJSONTyped(json, false);
}
exports.NamespaceRegistrationTypeEnumFromJSON = NamespaceRegistrationTypeEnumFromJSON;
function NamespaceRegistrationTypeEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.NamespaceRegistrationTypeEnumFromJSONTyped = NamespaceRegistrationTypeEnumFromJSONTyped;
function NamespaceRegistrationTypeEnumToJSON(value) {
    return value;
}
exports.NamespaceRegistrationTypeEnumToJSON = NamespaceRegistrationTypeEnumToJSON;
