// symbol-openapi-typescript-fetch-client/dist/models/RolesTypeEnum.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.RolesTypeEnumToJSON = exports.RolesTypeEnumFromJSONTyped = exports.RolesTypeEnumFromJSON = exports.RolesTypeEnum = void 0;
/**
 * Role of the node:
 * * 1 - Peer node.
 * * 2 - Api node.
 * * 3 - Dual node.
 * @export
 * @enum {string}
 */
var RolesTypeEnum;
(function (RolesTypeEnum) {
    RolesTypeEnum[RolesTypeEnum["NUMBER_1"] = 1] = "NUMBER_1";
    RolesTypeEnum[RolesTypeEnum["NUMBER_2"] = 2] = "NUMBER_2";
    RolesTypeEnum[RolesTypeEnum["NUMBER_3"] = 3] = "NUMBER_3";
})(RolesTypeEnum = exports.RolesTypeEnum || (exports.RolesTypeEnum = {}));
function RolesTypeEnumFromJSON(json) {
    return RolesTypeEnumFromJSONTyped(json, false);
}
exports.RolesTypeEnumFromJSON = RolesTypeEnumFromJSON;
function RolesTypeEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.RolesTypeEnumFromJSONTyped = RolesTypeEnumFromJSONTyped;
function RolesTypeEnumToJSON(value) {
    return value;
}
exports.RolesTypeEnumToJSON = RolesTypeEnumToJSON;
