// symbol-openapi-typescript-fetch-client/dist/apis/MultisigRoutesApi.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.MultisigRoutesApi = void 0;
const runtime = require("../runtime");
const models_1 = require("../models");
/**
 *
 */
class MultisigRoutesApi extends runtime.BaseAPI {
    /**
     * Returns the multisig account information.
     * Get multisig account information
     */
    getAccountMultisigRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            if (requestParameters.address === null || requestParameters.address === undefined) {
                throw new runtime.RequiredError('address', 'Required parameter requestParameters.address was null or undefined when calling getAccountMultisig.');
            }
            const queryParameters = {};
            const headerParameters = {};
            const response = yield this.request({
                path: `/account/{address}/multisig`.replace(`{${"address"}}`, encodeURIComponent(String(requestParameters.address))),
                method: 'GET',
                headers: headerParameters,
                query: queryParameters,
            });
            return new runtime.JSONApiResponse(response, (jsonValue) => models_1.MultisigAccountInfoDTOFromJSON(jsonValue));
        });
    }
    /**
     * Returns the multisig account information.
     * Get multisig account information
     */
    getAccountMultisig(address) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.getAccountMultisigRaw({ address: address });
            return yield response.value();
        });
    }
    /**
     * Returns the multisig account graph.
     * Get multisig account graph information
     */
    getAccountMultisigGraphRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            if (requestParameters.address === null || requestParameters.address === undefined) {
                throw new runtime.RequiredError('address', 'Required parameter requestParameters.address was null or undefined when calling getAccountMultisigGraph.');
            }
            const queryParameters = {};
            const headerParameters = {};
            const response = yield this.request({
                path: `/account/{address}/multisig/graph`.replace(`{${"address"}}`, encodeURIComponent(String(requestParameters.address))),
                method: 'GET',
                headers: headerParameters,
                query: queryParameters,
            });
            return new runtime.JSONApiResponse(response, (jsonValue) => jsonValue.map(models_1.MultisigAccountGraphInfoDTOFromJSON));
        });
    }
    /**
     * Returns the multisig account graph.
     * Get multisig account graph information
     */
    getAccountMultisigGraph(address) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.getAccountMultisigGraphRaw({ address: address });
            return yield response.value();
        });
    }
}
exports.MultisigRoutesApi = MultisigRoutesApi;
