// symbol-openapi-typescript-fetch-client/dist/apis/RestrictionMosaicRoutesApi.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.RestrictionMosaicRoutesApi = void 0;
const runtime = require("../runtime");
const models_1 = require("../models");
/**
 *
 */
class RestrictionMosaicRoutesApi extends runtime.BaseAPI {
    /**
     * Get mosaic address restriction.
     * Get mosaic address restrictions for a given mosaic and account address.
     */
    getMosaicAddressRestrictionRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            if (requestParameters.mosaicId === null || requestParameters.mosaicId === undefined) {
                throw new runtime.RequiredError('mosaicId', 'Required parameter requestParameters.mosaicId was null or undefined when calling getMosaicAddressRestriction.');
            }
            if (requestParameters.address === null || requestParameters.address === undefined) {
                throw new runtime.RequiredError('address', 'Required parameter requestParameters.address was null or undefined when calling getMosaicAddressRestriction.');
            }
            const queryParameters = {};
            const headerParameters = {};
            const response = yield this.request({
                path: `/restrictions/mosaic/{mosaicId}/address/{address}`.replace(`{${"mosaicId"}}`, encodeURIComponent(String(requestParameters.mosaicId))).replace(`{${"address"}}`, encodeURIComponent(String(requestParameters.address))),
                method: 'GET',
                headers: headerParameters,
                query: queryParameters,
            });
            return new runtime.JSONApiResponse(response, (jsonValue) => models_1.MosaicAddressRestrictionDTOFromJSON(jsonValue));
        });
    }
    /**
     * Get mosaic address restriction.
     * Get mosaic address restrictions for a given mosaic and account address.
     */
    getMosaicAddressRestriction(mosaicId, address) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.getMosaicAddressRestrictionRaw({ mosaicId: mosaicId, address: address });
            return yield response.value();
        });
    }
    /**
     * Get mosaic address restrictions.
     * Get mosaic address restrictions for a given mosaic and account identifiers array.
     */
    getMosaicAddressRestrictionsRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            if (requestParameters.mosaicId === null || requestParameters.mosaicId === undefined) {
                throw new runtime.RequiredError('mosaicId', 'Required parameter requestParameters.mosaicId was null or undefined when calling getMosaicAddressRestrictions.');
            }
            const queryParameters = {};
            const headerParameters = {};
            headerParameters['Content-Type'] = 'application/json';
            const response = yield this.request({
                path: `/restrictions/mosaic/{mosaicId}`.replace(`{${"mosaicId"}}`, encodeURIComponent(String(requestParameters.mosaicId))),
                method: 'POST',
                headers: headerParameters,
                query: queryParameters,
                body: models_1.AccountIdsToJSON(requestParameters.accountIds),
            });
            return new runtime.JSONApiResponse(response, (jsonValue) => jsonValue.map(models_1.MosaicAddressRestrictionDTOFromJSON));
        });
    }
    /**
     * Get mosaic address restrictions.
     * Get mosaic address restrictions for a given mosaic and account identifiers array.
     */
    getMosaicAddressRestrictions(mosaicId, accountIds) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.getMosaicAddressRestrictionsRaw({ mosaicId: mosaicId, accountIds: accountIds });
            return yield response.value();
        });
    }
    /**
     * Get mosaic global restriction.
     * Get mosaic global restriction for a given mosaic identifier.
     */
    getMosaicGlobalRestrictionRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            if (requestParameters.mosaicId === null || requestParameters.mosaicId === undefined) {
                throw new runtime.RequiredError('mosaicId', 'Required parameter requestParameters.mosaicId was null or undefined when calling getMosaicGlobalRestriction.');
            }
            const queryParameters = {};
            const headerParameters = {};
            const response = yield this.request({
                path: `/restrictions/mosaic/{mosaicId}`.replace(`{${"mosaicId"}}`, encodeURIComponent(String(requestParameters.mosaicId))),
                method: 'GET',
                headers: headerParameters,
                query: queryParameters,
            });
            return new runtime.JSONApiResponse(response, (jsonValue) => models_1.MosaicGlobalRestrictionDTOFromJSON(jsonValue));
        });
    }
    /**
     * Get mosaic global restriction.
     * Get mosaic global restriction for a given mosaic identifier.
     */
    getMosaicGlobalRestriction(mosaicId) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.getMosaicGlobalRestrictionRaw({ mosaicId: mosaicId });
            return yield response.value();
        });
    }
    /**
     * Get mosaic global restrictions.
     * Get mosaic global restrictions for an array of mosaics.
     */
    getMosaicGlobalRestrictionsRaw(requestParameters) {
        return __awaiter(this, void 0, void 0, function* () {
            if (requestParameters.mosaicIds === null || requestParameters.mosaicIds === undefined) {
                throw new runtime.RequiredError('mosaicIds', 'Required parameter requestParameters.mosaicIds was null or undefined when calling getMosaicGlobalRestrictions.');
            }
            const queryParameters = {};
            const headerParameters = {};
            headerParameters['Content-Type'] = 'application/json';
            const response = yield this.request({
                path: `/restrictions/mosaic`,
                method: 'POST',
                headers: headerParameters,
                query: queryParameters,
                body: models_1.MosaicIdsToJSON(requestParameters.mosaicIds),
            });
            return new runtime.JSONApiResponse(response, (jsonValue) => jsonValue.map(models_1.MosaicGlobalRestrictionDTOFromJSON));
        });
    }
    /**
     * Get mosaic global restrictions.
     * Get mosaic global restrictions for an array of mosaics.
     */
    getMosaicGlobalRestrictions(mosaicIds) {
        return __awaiter(this, void 0, void 0, function* () {
            const response = yield this.getMosaicGlobalRestrictionsRaw({ mosaicIds: mosaicIds });
            return yield response.value();
        });
    }
}
exports.RestrictionMosaicRoutesApi = RestrictionMosaicRoutesApi;
