// symbol-openapi-typescript-fetch-client/dist/models/MosaicRestrictionEntryTypeEnum.js
"use strict";
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 *
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */
Object.defineProperty(exports, "__esModule", { value: true });
exports.MosaicRestrictionEntryTypeEnumToJSON = exports.MosaicRestrictionEntryTypeEnumFromJSONTyped = exports.MosaicRestrictionEntryTypeEnumFromJSON = exports.MosaicRestrictionEntryTypeEnum = void 0;
/**
 * - 0 - Mosaic address restriction.
 * - 1 - Mosaic global restriction.
 * @export
 * @enum {string}
 */
var MosaicRestrictionEntryTypeEnum;
(function (MosaicRestrictionEntryTypeEnum) {
    MosaicRestrictionEntryTypeEnum[MosaicRestrictionEntryTypeEnum["NUMBER_0"] = 0] = "NUMBER_0";
    MosaicRestrictionEntryTypeEnum[MosaicRestrictionEntryTypeEnum["NUMBER_1"] = 1] = "NUMBER_1";
})(MosaicRestrictionEntryTypeEnum = exports.MosaicRestrictionEntryTypeEnum || (exports.MosaicRestrictionEntryTypeEnum = {}));
function MosaicRestrictionEntryTypeEnumFromJSON(json) {
    return MosaicRestrictionEntryTypeEnumFromJSONTyped(json, false);
}
exports.MosaicRestrictionEntryTypeEnumFromJSON = MosaicRestrictionEntryTypeEnumFromJSON;
function MosaicRestrictionEntryTypeEnumFromJSONTyped(json, ignoreDiscriminator) {
    return json;
}
exports.MosaicRestrictionEntryTypeEnumFromJSONTyped = MosaicRestrictionEntryTypeEnumFromJSONTyped;
function MosaicRestrictionEntryTypeEnumToJSON(value) {
    return value;
}
exports.MosaicRestrictionEntryTypeEnumToJSON = MosaicRestrictionEntryTypeEnumToJSON;
