// catbuffer-typescript/dist/AccountRestrictionsInfoBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountRestrictionsInfoBuilder = void 0;
const AccountRestrictionAddressValueBuilder_1 = require("./AccountRestrictionAddressValueBuilder");
const AccountRestrictionFlagsDto_1 = require("./AccountRestrictionFlagsDto");
const AccountRestrictionMosaicValueBuilder_1 = require("./AccountRestrictionMosaicValueBuilder");
const AccountRestrictionTransactionTypeValueBuilder_1 = require("./AccountRestrictionTransactionTypeValueBuilder");
const GeneratorUtils_1 = require("./GeneratorUtils");
class AccountRestrictionsInfoBuilder {
    constructor(restrictionFlags, addressRestrictions, mosaicIdRestrictions, transactionTypeRestrictions) {
        this.restrictionFlags = restrictionFlags;
        this.addressRestrictions = addressRestrictions;
        this.mosaicIdRestrictions = mosaicIdRestrictions;
        this.transactionTypeRestrictions = transactionTypeRestrictions;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const restrictionFlags = GeneratorUtils_1.GeneratorUtils.bufferToUint(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 2));
        byteArray.splice(0, 2);
        const restrictionFlagsConditionBytes = Uint8Array.from(byteArray.slice(0, 1));
        byteArray.splice(0, 1);
        let addressRestrictions;
        if (restrictionFlags & AccountRestrictionFlagsDto_1.AccountRestrictionFlagsDto.ADDRESS) {
            addressRestrictions = AccountRestrictionAddressValueBuilder_1.AccountRestrictionAddressValueBuilder.loadFromBinary(restrictionFlagsConditionBytes);
        }
        let mosaicIdRestrictions;
        if (restrictionFlags & AccountRestrictionFlagsDto_1.AccountRestrictionFlagsDto.MOSAIC_ID) {
            mosaicIdRestrictions = AccountRestrictionMosaicValueBuilder_1.AccountRestrictionMosaicValueBuilder.loadFromBinary(restrictionFlagsConditionBytes);
        }
        let transactionTypeRestrictions;
        if (restrictionFlags & AccountRestrictionFlagsDto_1.AccountRestrictionFlagsDto.TRANSACTION_TYPE) {
            transactionTypeRestrictions = AccountRestrictionTransactionTypeValueBuilder_1.AccountRestrictionTransactionTypeValueBuilder.loadFromBinary(restrictionFlagsConditionBytes);
        }
        return new AccountRestrictionsInfoBuilder(restrictionFlags, addressRestrictions, mosaicIdRestrictions, transactionTypeRestrictions);
    }
    getRestrictionFlags() {
        return this.restrictionFlags;
    }
    getAddressRestrictions() {
        if (this.restrictionFlags & AccountRestrictionFlagsDto_1.AccountRestrictionFlagsDto.ADDRESS) {
            throw new Error('restrictionFlags is not set to ADDRESS.');
        }
        return this.addressRestrictions;
    }
    getMosaicIdRestrictions() {
        if (this.restrictionFlags & AccountRestrictionFlagsDto_1.AccountRestrictionFlagsDto.MOSAIC_ID) {
            throw new Error('restrictionFlags is not set to MOSAIC_ID.');
        }
        return this.mosaicIdRestrictions;
    }
    getTransactionTypeRestrictions() {
        if (this.restrictionFlags & AccountRestrictionFlagsDto_1.AccountRestrictionFlagsDto.TRANSACTION_TYPE) {
            throw new Error('restrictionFlags is not set to TRANSACTION_TYPE.');
        }
        return this.transactionTypeRestrictions;
    }
    getSize() {
        let size = 0;
        size += 2;
        if (this.restrictionFlags & AccountRestrictionFlagsDto_1.AccountRestrictionFlagsDto.ADDRESS) {
            size += this.addressRestrictions.getSize();
        }
        if (this.restrictionFlags & AccountRestrictionFlagsDto_1.AccountRestrictionFlagsDto.MOSAIC_ID) {
            size += this.mosaicIdRestrictions.getSize();
        }
        if (this.restrictionFlags & AccountRestrictionFlagsDto_1.AccountRestrictionFlagsDto.TRANSACTION_TYPE) {
            size += this.transactionTypeRestrictions.getSize();
        }
        return size;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const restrictionFlagsBytes = GeneratorUtils_1.GeneratorUtils.uintToBuffer(this.getRestrictionFlags(), 2);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, restrictionFlagsBytes);
        if (this.restrictionFlags & AccountRestrictionFlagsDto_1.AccountRestrictionFlagsDto.ADDRESS) {
            const addressRestrictionsBytes = this.addressRestrictions.serialize();
            newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, addressRestrictionsBytes);
        }
        if (this.restrictionFlags & AccountRestrictionFlagsDto_1.AccountRestrictionFlagsDto.MOSAIC_ID) {
            const mosaicIdRestrictionsBytes = this.mosaicIdRestrictions.serialize();
            newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, mosaicIdRestrictionsBytes);
        }
        if (this.restrictionFlags & AccountRestrictionFlagsDto_1.AccountRestrictionFlagsDto.TRANSACTION_TYPE) {
            const transactionTypeRestrictionsBytes = this.transactionTypeRestrictions.serialize();
            newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, transactionTypeRestrictionsBytes);
        }
        return newArray;
    }
}
exports.AccountRestrictionsInfoBuilder = AccountRestrictionsInfoBuilder;
