// catbuffer-typescript/dist/AccountRestrictionTransactionTypeValueBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.AccountRestrictionTransactionTypeValueBuilder = void 0;
const GeneratorUtils_1 = require("./GeneratorUtils");
class AccountRestrictionTransactionTypeValueBuilder {
    constructor(restrictionValues) {
        this.restrictionValues = restrictionValues;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const restrictionValuesCount = GeneratorUtils_1.GeneratorUtils.bufferToUint64(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 8));
        byteArray.splice(0, 8);
        const restrictionValues = [];
        for (let i = 0; i < (Array.isArray(restrictionValuesCount) ? GeneratorUtils_1.GeneratorUtils.compact(restrictionValuesCount) : restrictionValuesCount); i++) {
            const item = GeneratorUtils_1.GeneratorUtils.bufferToUint(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 2));
            restrictionValues.push(item);
            byteArray.splice(0, 2);
        }
        return new AccountRestrictionTransactionTypeValueBuilder(restrictionValues);
    }
    getRestrictionValues() {
        return this.restrictionValues;
    }
    getSize() {
        let size = 0;
        size += 8;
        this.restrictionValues.forEach(() => size += 2);
        return size;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const restrictionValuesCountBytes = GeneratorUtils_1.GeneratorUtils.uint64ToBuffer(GeneratorUtils_1.GeneratorUtils.fromUint(this.restrictionValues.length));
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, restrictionValuesCountBytes);
        this.restrictionValues.forEach((item) => {
            const restrictionValuesBytes = GeneratorUtils_1.GeneratorUtils.uintToBuffer(item, 2);
            newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, restrictionValuesBytes);
        });
        return newArray;
    }
}
exports.AccountRestrictionTransactionTypeValueBuilder = AccountRestrictionTransactionTypeValueBuilder;
