// catbuffer-typescript/dist/BlockDurationDto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockDurationDto = void 0;
const GeneratorUtils_1 = require("./GeneratorUtils");
class BlockDurationDto {
    constructor(blockDuration) {
        this.blockDuration = blockDuration;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const blockDuration = GeneratorUtils_1.GeneratorUtils.bufferToUint64(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 8));
        byteArray.splice(0, 8);
        return new BlockDurationDto(blockDuration);
    }
    getBlockDuration() {
        return this.blockDuration;
    }
    getSize() {
        return 8;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const blockDurationBytes = GeneratorUtils_1.GeneratorUtils.uint64ToBuffer(this.getBlockDuration());
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, blockDurationBytes);
        return newArray;
    }
}
exports.BlockDurationDto = BlockDurationDto;
