// catbuffer-typescript/dist/DifficultyDto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.DifficultyDto = void 0;
const GeneratorUtils_1 = require("./GeneratorUtils");
class DifficultyDto {
    constructor(difficulty) {
        this.difficulty = difficulty;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const difficulty = GeneratorUtils_1.GeneratorUtils.bufferToUint64(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 8));
        byteArray.splice(0, 8);
        return new DifficultyDto(difficulty);
    }
    getDifficulty() {
        return this.difficulty;
    }
    getSize() {
        return 8;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const difficultyBytes = GeneratorUtils_1.GeneratorUtils.uint64ToBuffer(this.getDifficulty());
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, difficultyBytes);
        return newArray;
    }
}
exports.DifficultyDto = DifficultyDto;
