// catbuffer-typescript/dist/FinalizationEpochDto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FinalizationEpochDto = void 0;
const GeneratorUtils_1 = require("./GeneratorUtils");
class FinalizationEpochDto {
    constructor(finalizationEpoch) {
        this.finalizationEpoch = finalizationEpoch;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const finalizationEpoch = GeneratorUtils_1.GeneratorUtils.bufferToUint(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 4));
        byteArray.splice(0, 4);
        return new FinalizationEpochDto(finalizationEpoch);
    }
    getFinalizationEpoch() {
        return this.finalizationEpoch;
    }
    getSize() {
        return 4;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const finalizationEpochBytes = GeneratorUtils_1.GeneratorUtils.uintToBuffer(this.getFinalizationEpoch(), 4);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, finalizationEpochBytes);
        return newArray;
    }
}
exports.FinalizationEpochDto = FinalizationEpochDto;
