// catbuffer-typescript/dist/ImportanceHeightDto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImportanceHeightDto = void 0;
const GeneratorUtils_1 = require("./GeneratorUtils");
class ImportanceHeightDto {
    constructor(importanceHeight) {
        this.importanceHeight = importanceHeight;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const importanceHeight = GeneratorUtils_1.GeneratorUtils.bufferToUint64(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 8));
        byteArray.splice(0, 8);
        return new ImportanceHeightDto(importanceHeight);
    }
    getImportanceHeight() {
        return this.importanceHeight;
    }
    getSize() {
        return 8;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const importanceHeightBytes = GeneratorUtils_1.GeneratorUtils.uint64ToBuffer(this.getImportanceHeight());
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, importanceHeightBytes);
        return newArray;
    }
}
exports.ImportanceHeightDto = ImportanceHeightDto;
