// catbuffer-typescript/dist/MosaicRestrictionKeyDto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MosaicRestrictionKeyDto = void 0;
const GeneratorUtils_1 = require("./GeneratorUtils");
class MosaicRestrictionKeyDto {
    constructor(mosaicRestrictionKey) {
        this.mosaicRestrictionKey = mosaicRestrictionKey;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const mosaicRestrictionKey = GeneratorUtils_1.GeneratorUtils.bufferToUint64(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 8));
        byteArray.splice(0, 8);
        return new MosaicRestrictionKeyDto(mosaicRestrictionKey);
    }
    getMosaicRestrictionKey() {
        return this.mosaicRestrictionKey;
    }
    getSize() {
        return 8;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const mosaicRestrictionKeyBytes = GeneratorUtils_1.GeneratorUtils.uint64ToBuffer(this.getMosaicRestrictionKey());
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, mosaicRestrictionKeyBytes);
        return newArray;
    }
}
exports.MosaicRestrictionKeyDto = MosaicRestrictionKeyDto;
