// catbuffer-typescript/dist/ScopedMetadataKeyDto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ScopedMetadataKeyDto = void 0;
const GeneratorUtils_1 = require("./GeneratorUtils");
class ScopedMetadataKeyDto {
    constructor(scopedMetadataKey) {
        this.scopedMetadataKey = scopedMetadataKey;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const scopedMetadataKey = GeneratorUtils_1.GeneratorUtils.bufferToUint64(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 8));
        byteArray.splice(0, 8);
        return new ScopedMetadataKeyDto(scopedMetadataKey);
    }
    getScopedMetadataKey() {
        return this.scopedMetadataKey;
    }
    getSize() {
        return 8;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const scopedMetadataKeyBytes = GeneratorUtils_1.GeneratorUtils.uint64ToBuffer(this.getScopedMetadataKey());
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, scopedMetadataKeyBytes);
        return newArray;
    }
}
exports.ScopedMetadataKeyDto = ScopedMetadataKeyDto;
