// catbuffer-typescript/src/AddressKeyValueBuilder.ts
// tslint:disable: jsdoc-format
/**
*** Copyright (c) 2016-present,
*** Jaguar0625, gimre, BloodyRookie, Tech Bureau, Corp. All rights reserved.
***
*** This file is part of Catapult.
***
*** Catapult is free software: you can redistribute it and/or modify
*** it under the terms of the GNU Lesser General Public License as published by
*** the Free Software Foundation, either version 3 of the License, or
*** (at your option) any later version.
***
*** Catapult is distributed in the hope that it will be useful,
*** but WITHOUT ANY WARRANTY; without even the implied warranty of
*** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
*** GNU Lesser General Public License for more details.
***
*** You should have received a copy of the GNU Lesser General Public License
*** along with Catapult. If not, see <http://www.gnu.org/licenses/>.
**/

import { GeneratorUtils } from './GeneratorUtils';
import { MosaicRestrictionKeyDto } from './MosaicRestrictionKeyDto';

/** Layout for mosaic address restriction key-value pair. */
export class AddressKeyValueBuilder {
    /** Key for value. */
    key: MosaicRestrictionKeyDto;
    /** Value associated by key. */
    value: number[];

    /**
     * Constructor.
     *
     * @param key Key for value.
     * @param value Value associated by key.
     */
    public constructor(key: MosaicRestrictionKeyDto,  value: number[]) {
        this.key = key;
        this.value = value;
    }

    /**
     * Creates an instance of AddressKeyValueBuilder from binary payload.
     *
     * @param payload Byte payload to use to serialize the object.
     * @return Instance of AddressKeyValueBuilder.
     */
    public static loadFromBinary(payload: Uint8Array): AddressKeyValueBuilder {
        const byteArray = Array.from(payload);
        const key = MosaicRestrictionKeyDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, key.getSize());
        const value = GeneratorUtils.bufferToUint64(GeneratorUtils.getBytes(Uint8Array.from(byteArray), 8));
        byteArray.splice(0, 8);
        return new AddressKeyValueBuilder(key, value);
    }

    /**
     * Gets key for value.
     *
     * @return Key for value.
     */
    public getKey(): MosaicRestrictionKeyDto {
        return this.key;
    }

    /**
     * Gets value associated by key.
     *
     * @return Value associated by key.
     */
    public getValue(): number[] {
        return this.value;
    }

    /**
     * Gets the size of the object.
     *
     * @return Size in bytes.
     */
    public getSize(): number {
        let size = 0;
        size += this.key.getSize();
        size += 8; // value
        return size;
    }

    /**
     * Serializes an object to bytes.
     *
     * @return Serialized bytes.
     */
    public serialize(): Uint8Array {
        let newArray = Uint8Array.from([]);
        const keyBytes = this.key.serialize();
        newArray = GeneratorUtils.concatTypedArrays(newArray, keyBytes);
        const valueBytes = GeneratorUtils.uint64ToBuffer(this.getValue());
        newArray = GeneratorUtils.concatTypedArrays(newArray, valueBytes);
        return newArray;
    }
}
