// catbuffer-typescript/src/HeightActivityBucketsBuilder.ts
// tslint:disable: jsdoc-format
/**
*** Copyright (c) 2016-present,
*** Jaguar0625, gimre, BloodyRookie, Tech Bureau, Corp. All rights reserved.
***
*** This file is part of Catapult.
***
*** Catapult is free software: you can redistribute it and/or modify
*** it under the terms of the GNU Lesser General Public License as published by
*** the Free Software Foundation, either version 3 of the License, or
*** (at your option) any later version.
***
*** Catapult is distributed in the hope that it will be useful,
*** but WITHOUT ANY WARRANTY; without even the implied warranty of
*** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
*** GNU Lesser General Public License for more details.
***
*** You should have received a copy of the GNU Lesser General Public License
*** along with Catapult. If not, see <http://www.gnu.org/licenses/>.
**/

import { GeneratorUtils } from './GeneratorUtils';

/** Account activity buckets. */
export class HeightActivityBucketsBuilder {
    /** Account activity buckets. */
    buckets: Uint8Array;

    /**
     * Constructor.
     *
     * @param buckets Account activity buckets.
     */
    public constructor(buckets: Uint8Array) {
        this.buckets = buckets;
    }

    /**
     * Creates an instance of HeightActivityBucketsBuilder from binary payload.
     *
     * @param payload Byte payload to use to serialize the object.
     * @return Instance of HeightActivityBucketsBuilder.
     */
    public static loadFromBinary(payload: Uint8Array): HeightActivityBucketsBuilder {
        const byteArray = Array.from(payload);
        const buckets = GeneratorUtils.getBytes(Uint8Array.from(byteArray), 5);
        byteArray.splice(0, 5);
        return new HeightActivityBucketsBuilder(buckets);
    }

    /**
     * Gets account activity buckets.
     *
     * @return Account activity buckets.
     */
    public getBuckets(): Uint8Array {
        return this.buckets;
    }

    /**
     * Gets the size of the object.
     *
     * @return Size in bytes.
     */
    public getSize(): number {
        let size = 0;
        size += this.buckets.length;
        return size;
    }

    /**
     * Serializes an object to bytes.
     *
     * @return Serialized bytes.
     */
    public serialize(): Uint8Array {
        let newArray = Uint8Array.from([]);
        newArray = GeneratorUtils.concatTypedArrays(newArray, this.buckets);
        return newArray;
    }
}
