// catbuffer-typescript/src/MultisigAccountModificationTransactionBodyBuilder.ts
// tslint:disable: jsdoc-format
/**
*** Copyright (c) 2016-present,
*** Jaguar0625, gimre, BloodyRookie, Tech Bureau, Corp. All rights reserved.
***
*** This file is part of Catapult.
***
*** Catapult is free software: you can redistribute it and/or modify
*** it under the terms of the GNU Lesser General Public License as published by
*** the Free Software Foundation, either version 3 of the License, or
*** (at your option) any later version.
***
*** Catapult is distributed in the hope that it will be useful,
*** but WITHOUT ANY WARRANTY; without even the implied warranty of
*** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
*** GNU Lesser General Public License for more details.
***
*** You should have received a copy of the GNU Lesser General Public License
*** along with Catapult. If not, see <http://www.gnu.org/licenses/>.
**/

import { GeneratorUtils } from './GeneratorUtils';
import { UnresolvedAddressDto } from './UnresolvedAddressDto';

/** Binary layout for a multisig account modification transaction. */
export class MultisigAccountModificationTransactionBodyBuilder {
    /** Relative change of the minimal number of cosignatories required when removing an account. */
    minRemovalDelta: number;
    /** Relative change of the minimal number of cosignatories required when approving a transaction. */
    minApprovalDelta: number;
    /** Reserved padding to align addressAdditions on 8-byte boundary. */
    multisigAccountModificationTransactionBody_Reserved1: number;
    /** Cosignatory address additions. */
    addressAdditions: UnresolvedAddressDto[];
    /** Cosignatory address deletions. */
    addressDeletions: UnresolvedAddressDto[];

    /**
     * Constructor.
     *
     * @param minRemovalDelta Relative change of the minimal number of cosignatories required when removing an account.
     * @param minApprovalDelta Relative change of the minimal number of cosignatories required when approving a transaction.
     * @param addressAdditions Cosignatory address additions.
     * @param addressDeletions Cosignatory address deletions.
     */
    // tslint:disable-next-line: max-line-length
    public constructor(minRemovalDelta: number,  minApprovalDelta: number,  addressAdditions: UnresolvedAddressDto[],  addressDeletions: UnresolvedAddressDto[]) {
        this.minRemovalDelta = minRemovalDelta;
        this.minApprovalDelta = minApprovalDelta;
        this.multisigAccountModificationTransactionBody_Reserved1 = 0;
        this.addressAdditions = addressAdditions;
        this.addressDeletions = addressDeletions;
    }

    /**
     * Creates an instance of MultisigAccountModificationTransactionBodyBuilder from binary payload.
     *
     * @param payload Byte payload to use to serialize the object.
     * @return Instance of MultisigAccountModificationTransactionBodyBuilder.
     */
    public static loadFromBinary(payload: Uint8Array): MultisigAccountModificationTransactionBodyBuilder {
        const byteArray = Array.from(payload);
        const minRemovalDelta = GeneratorUtils.bufferToUint(GeneratorUtils.getBytes(Uint8Array.from(byteArray), 1));
        byteArray.splice(0, 1);
        const minApprovalDelta = GeneratorUtils.bufferToUint(GeneratorUtils.getBytes(Uint8Array.from(byteArray), 1));
        byteArray.splice(0, 1);
        const addressAdditionsCount = GeneratorUtils.bufferToUint(GeneratorUtils.getBytes(Uint8Array.from(byteArray), 1));
        byteArray.splice(0, 1);
        const addressDeletionsCount = GeneratorUtils.bufferToUint(GeneratorUtils.getBytes(Uint8Array.from(byteArray), 1));
        byteArray.splice(0, 1);
        // tslint:disable-next-line: max-line-length
        const multisigAccountModificationTransactionBody_Reserved1 = GeneratorUtils.bufferToUint(GeneratorUtils.getBytes(Uint8Array.from(byteArray), 4));
        byteArray.splice(0, 4);
        const addressAdditions: UnresolvedAddressDto[] = [];
        // tslint:disable-next-line: max-line-length
        for (let i = 0; i < (Array.isArray(addressAdditionsCount) ? GeneratorUtils.compact(addressAdditionsCount) : addressAdditionsCount); i++) {
            const item = UnresolvedAddressDto.loadFromBinary(Uint8Array.from(byteArray));
            addressAdditions.push(item);
            byteArray.splice(0, item.getSize());
        }
        const addressDeletions: UnresolvedAddressDto[] = [];
        // tslint:disable-next-line: max-line-length
        for (let i = 0; i < (Array.isArray(addressDeletionsCount) ? GeneratorUtils.compact(addressDeletionsCount) : addressDeletionsCount); i++) {
            const item = UnresolvedAddressDto.loadFromBinary(Uint8Array.from(byteArray));
            addressDeletions.push(item);
            byteArray.splice(0, item.getSize());
        }
        // tslint:disable-next-line: max-line-length
        return new MultisigAccountModificationTransactionBodyBuilder(minRemovalDelta, minApprovalDelta, addressAdditions, addressDeletions);
    }

    /**
     * Gets relative change of the minimal number of cosignatories required when removing an account.
     *
     * @return Relative change of the minimal number of cosignatories required when removing an account.
     */
    public getMinRemovalDelta(): number {
        return this.minRemovalDelta;
    }

    /**
     * Gets relative change of the minimal number of cosignatories required when approving a transaction.
     *
     * @return Relative change of the minimal number of cosignatories required when approving a transaction.
     */
    public getMinApprovalDelta(): number {
        return this.minApprovalDelta;
    }

    /**
     * Gets reserved padding to align addressAdditions on 8-byte boundary.
     *
     * @return Reserved padding to align addressAdditions on 8-byte boundary.
     */
    public getMultisigAccountModificationTransactionBody_Reserved1(): number {
        return this.multisigAccountModificationTransactionBody_Reserved1;
    }

    /**
     * Gets cosignatory address additions.
     *
     * @return Cosignatory address additions.
     */
    public getAddressAdditions(): UnresolvedAddressDto[] {
        return this.addressAdditions;
    }

    /**
     * Gets cosignatory address deletions.
     *
     * @return Cosignatory address deletions.
     */
    public getAddressDeletions(): UnresolvedAddressDto[] {
        return this.addressDeletions;
    }

    /**
     * Gets the size of the object.
     *
     * @return Size in bytes.
     */
    public getSize(): number {
        let size = 0;
        size += 1; // minRemovalDelta
        size += 1; // minApprovalDelta
        size += 1; // addressAdditionsCount
        size += 1; // addressDeletionsCount
        size += 4; // multisigAccountModificationTransactionBody_Reserved1
        this.addressAdditions.forEach((o) => size += o.getSize());
        this.addressDeletions.forEach((o) => size += o.getSize());
        return size;
    }

    /**
     * Serializes an object to bytes.
     *
     * @return Serialized bytes.
     */
    public serialize(): Uint8Array {
        let newArray = Uint8Array.from([]);
        const minRemovalDeltaBytes = GeneratorUtils.uintToBuffer(this.getMinRemovalDelta(), 1);
        newArray = GeneratorUtils.concatTypedArrays(newArray, minRemovalDeltaBytes);
        const minApprovalDeltaBytes = GeneratorUtils.uintToBuffer(this.getMinApprovalDelta(), 1);
        newArray = GeneratorUtils.concatTypedArrays(newArray, minApprovalDeltaBytes);
        const addressAdditionsCountBytes = GeneratorUtils.uintToBuffer(this.addressAdditions.length, 1);
        newArray = GeneratorUtils.concatTypedArrays(newArray, addressAdditionsCountBytes);
        const addressDeletionsCountBytes = GeneratorUtils.uintToBuffer(this.addressDeletions.length, 1);
        newArray = GeneratorUtils.concatTypedArrays(newArray, addressDeletionsCountBytes);
        // tslint:disable-next-line: max-line-length
        const multisigAccountModificationTransactionBody_Reserved1Bytes = GeneratorUtils.uintToBuffer(this.getMultisigAccountModificationTransactionBody_Reserved1(), 4);
        newArray = GeneratorUtils.concatTypedArrays(newArray, multisigAccountModificationTransactionBody_Reserved1Bytes);
        this.addressAdditions.forEach((item) => {
            const addressAdditionsBytes = item.serialize();
            newArray = GeneratorUtils.concatTypedArrays(newArray, addressAdditionsBytes);
        });
        this.addressDeletions.forEach((item) => {
            const addressDeletionsBytes = item.serialize();
            newArray = GeneratorUtils.concatTypedArrays(newArray, addressDeletionsBytes);
        });
        return newArray;
    }
}
