// catbuffer-typescript/src/VotingKeyDto.ts
// tslint:disable: jsdoc-format
/**
*** Copyright (c) 2016-present,
*** Jaguar0625, gimre, BloodyRookie, Tech Bureau, Corp. All rights reserved.
***
*** This file is part of Catapult.
***
*** Catapult is free software: you can redistribute it and/or modify
*** it under the terms of the GNU Lesser General Public License as published by
*** the Free Software Foundation, either version 3 of the License, or
*** (at your option) any later version.
***
*** Catapult is distributed in the hope that it will be useful,
*** but WITHOUT ANY WARRANTY; without even the implied warranty of
*** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
*** GNU Lesser General Public License for more details.
***
*** You should have received a copy of the GNU Lesser General Public License
*** along with Catapult. If not, see <http://www.gnu.org/licenses/>.
**/

import { GeneratorUtils } from './GeneratorUtils';

/** Voting key. */
export class VotingKeyDto {
    /** Voting key. */
    votingKey: Uint8Array;

    /**
     * Constructor.
     *
     * @param votingKey Voting key.
     */
     constructor(votingKey: Uint8Array) {
        this.votingKey = votingKey;
    }

    /**
     * Creates an instance of VotingKeyDto from binary payload.
     *
     * @param payload Byte payload to use to serialize the object.
     * @return Instance of VotingKeyDto.
     */
    public static loadFromBinary(payload: Uint8Array): VotingKeyDto {
        const byteArray = Array.from(payload);
        const votingKey = GeneratorUtils.getBytes(Uint8Array.from(byteArray), 48);
        byteArray.splice(0, 48);
        return new VotingKeyDto(votingKey);
    }

    /**
     * Gets Voting key.
     *
     * @return Voting key.
     */
    public getVotingKey(): Uint8Array {
        return this.votingKey;
    }

    /**
     * Gets the size of the object.
     *
     * @return Size in bytes.
     */
    public getSize(): number {
        return 48;
    }

    /**
     * Serializes an object to bytes.
     *
     * @return Serialized bytes.
     */
    public serialize(): Uint8Array {
        let newArray = Uint8Array.from([]);
        newArray = GeneratorUtils.concatTypedArrays(newArray, this.votingKey);
        return newArray;
    }
}
