// catbuffer-typescript/dist/BlockFeeMultiplierDto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.BlockFeeMultiplierDto = void 0;
const GeneratorUtils_1 = require("./GeneratorUtils");
class BlockFeeMultiplierDto {
    constructor(blockFeeMultiplier) {
        this.blockFeeMultiplier = blockFeeMultiplier;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const blockFeeMultiplier = GeneratorUtils_1.GeneratorUtils.bufferToUint(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 4));
        byteArray.splice(0, 4);
        return new BlockFeeMultiplierDto(blockFeeMultiplier);
    }
    getBlockFeeMultiplier() {
        return this.blockFeeMultiplier;
    }
    getSize() {
        return 4;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const blockFeeMultiplierBytes = GeneratorUtils_1.GeneratorUtils.uintToBuffer(this.getBlockFeeMultiplier(), 4);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, blockFeeMultiplierBytes);
        return newArray;
    }
}
exports.BlockFeeMultiplierDto = BlockFeeMultiplierDto;
