// catbuffer-typescript/dist/FinalizationPointDto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.FinalizationPointDto = void 0;
const GeneratorUtils_1 = require("./GeneratorUtils");
class FinalizationPointDto {
    constructor(finalizationPoint) {
        this.finalizationPoint = finalizationPoint;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const finalizationPoint = GeneratorUtils_1.GeneratorUtils.bufferToUint64(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 8));
        byteArray.splice(0, 8);
        return new FinalizationPointDto(finalizationPoint);
    }
    getFinalizationPoint() {
        return this.finalizationPoint;
    }
    getSize() {
        return 8;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const finalizationPointBytes = GeneratorUtils_1.GeneratorUtils.uint64ToBuffer(this.getFinalizationPoint());
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, finalizationPointBytes);
        return newArray;
    }
}
exports.FinalizationPointDto = FinalizationPointDto;
