// catbuffer-typescript/dist/ImportanceDto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ImportanceDto = void 0;
const GeneratorUtils_1 = require("./GeneratorUtils");
class ImportanceDto {
    constructor(importance) {
        this.importance = importance;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const importance = GeneratorUtils_1.GeneratorUtils.bufferToUint64(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 8));
        byteArray.splice(0, 8);
        return new ImportanceDto(importance);
    }
    getImportance() {
        return this.importance;
    }
    getSize() {
        return 8;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const importanceBytes = GeneratorUtils_1.GeneratorUtils.uint64ToBuffer(this.getImportance());
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, importanceBytes);
        return newArray;
    }
}
exports.ImportanceDto = ImportanceDto;
