// catbuffer-typescript/dist/MetadataValueBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MetadataValueBuilder = void 0;
const GeneratorUtils_1 = require("./GeneratorUtils");
class MetadataValueBuilder {
    constructor(data) {
        this.size = 0;
        this.data = data;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const size = GeneratorUtils_1.GeneratorUtils.bufferToUint(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 2));
        byteArray.splice(0, 2);
        const data = GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), size);
        byteArray.splice(0, size);
        return new MetadataValueBuilder(data);
    }
    getData() {
        return this.data;
    }
    getSize() {
        let size = 0;
        size += 2;
        size += this.data.length;
        return size;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const sizeBytes = GeneratorUtils_1.GeneratorUtils.uintToBuffer(this.getSize(), 2);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, sizeBytes);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, this.data);
        return newArray;
    }
}
exports.MetadataValueBuilder = MetadataValueBuilder;
