// catbuffer-typescript/dist/MosaicNonceDto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MosaicNonceDto = void 0;
const GeneratorUtils_1 = require("./GeneratorUtils");
class MosaicNonceDto {
    constructor(mosaicNonce) {
        this.mosaicNonce = mosaicNonce;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const mosaicNonce = GeneratorUtils_1.GeneratorUtils.bufferToUint(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 4));
        byteArray.splice(0, 4);
        return new MosaicNonceDto(mosaicNonce);
    }
    getMosaicNonce() {
        return this.mosaicNonce;
    }
    getSize() {
        return 4;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const mosaicNonceBytes = GeneratorUtils_1.GeneratorUtils.uintToBuffer(this.getMosaicNonce(), 4);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, mosaicNonceBytes);
        return newArray;
    }
}
exports.MosaicNonceDto = MosaicNonceDto;
