// catbuffer-typescript/dist/NamespaceIdDto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NamespaceIdDto = void 0;
const GeneratorUtils_1 = require("./GeneratorUtils");
class NamespaceIdDto {
    constructor(namespaceId) {
        this.namespaceId = namespaceId;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const namespaceId = GeneratorUtils_1.GeneratorUtils.bufferToUint64(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 8));
        byteArray.splice(0, 8);
        return new NamespaceIdDto(namespaceId);
    }
    getNamespaceId() {
        return this.namespaceId;
    }
    getSize() {
        return 8;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const namespaceIdBytes = GeneratorUtils_1.GeneratorUtils.uint64ToBuffer(this.getNamespaceId());
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, namespaceIdBytes);
        return newArray;
    }
}
exports.NamespaceIdDto = NamespaceIdDto;
