// catbuffer-typescript/dist/NamespacePathBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.NamespacePathBuilder = void 0;
const GeneratorUtils_1 = require("./GeneratorUtils");
const NamespaceAliasBuilder_1 = require("./NamespaceAliasBuilder");
class NamespacePathBuilder {
    constructor(path, alias) {
        this.path = path;
        this.alias = alias;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const pathSize = GeneratorUtils_1.GeneratorUtils.bufferToUint(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 1));
        byteArray.splice(0, 1);
        const path = GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), pathSize);
        byteArray.splice(0, pathSize);
        const alias = NamespaceAliasBuilder_1.NamespaceAliasBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, alias.getSize());
        return new NamespacePathBuilder(path, alias);
    }
    getPath() {
        return this.path;
    }
    getAlias() {
        return this.alias;
    }
    getSize() {
        let size = 0;
        size += 1;
        size += this.path.length;
        size += this.alias.getSize();
        return size;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const pathSizeBytes = GeneratorUtils_1.GeneratorUtils.uintToBuffer(this.path.length, 1);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, pathSizeBytes);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, this.path);
        const aliasBytes = this.alias.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, aliasBytes);
        return newArray;
    }
}
exports.NamespacePathBuilder = NamespacePathBuilder;
