// catbuffer-typescript/dist/PinnedVotingKeyBuilder.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.PinnedVotingKeyBuilder = void 0;
const FinalizationPointDto_1 = require("./FinalizationPointDto");
const GeneratorUtils_1 = require("./GeneratorUtils");
const VotingKeyDto_1 = require("./VotingKeyDto");
class PinnedVotingKeyBuilder {
    constructor(votingKey, startPoint, endPoint) {
        this.votingKey = votingKey;
        this.startPoint = startPoint;
        this.endPoint = endPoint;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const votingKey = VotingKeyDto_1.VotingKeyDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, votingKey.getSize());
        const startPoint = FinalizationPointDto_1.FinalizationPointDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, startPoint.getSize());
        const endPoint = FinalizationPointDto_1.FinalizationPointDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, endPoint.getSize());
        return new PinnedVotingKeyBuilder(votingKey, startPoint, endPoint);
    }
    getVotingKey() {
        return this.votingKey;
    }
    getStartPoint() {
        return this.startPoint;
    }
    getEndPoint() {
        return this.endPoint;
    }
    getSize() {
        let size = 0;
        size += this.votingKey.getSize();
        size += this.startPoint.getSize();
        size += this.endPoint.getSize();
        return size;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const votingKeyBytes = this.votingKey.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, votingKeyBytes);
        const startPointBytes = this.startPoint.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, startPointBytes);
        const endPointBytes = this.endPoint.serialize();
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, endPointBytes);
        return newArray;
    }
}
exports.PinnedVotingKeyBuilder = PinnedVotingKeyBuilder;
