// catbuffer-typescript/dist/ProofVerificationHashDto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ProofVerificationHashDto = void 0;
const GeneratorUtils_1 = require("./GeneratorUtils");
class ProofVerificationHashDto {
    constructor(proofVerificationHash) {
        this.proofVerificationHash = proofVerificationHash;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const proofVerificationHash = GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 16);
        byteArray.splice(0, 16);
        return new ProofVerificationHashDto(proofVerificationHash);
    }
    getProofVerificationHash() {
        return this.proofVerificationHash;
    }
    getSize() {
        return 16;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, this.proofVerificationHash);
        return newArray;
    }
}
exports.ProofVerificationHashDto = ProofVerificationHashDto;
