// catbuffer-typescript/dist/UnresolvedMosaicIdDto.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.UnresolvedMosaicIdDto = void 0;
const GeneratorUtils_1 = require("./GeneratorUtils");
class UnresolvedMosaicIdDto {
    constructor(unresolvedMosaicId) {
        this.unresolvedMosaicId = unresolvedMosaicId;
    }
    static loadFromBinary(payload) {
        const byteArray = Array.from(payload);
        const unresolvedMosaicId = GeneratorUtils_1.GeneratorUtils.bufferToUint64(GeneratorUtils_1.GeneratorUtils.getBytes(Uint8Array.from(byteArray), 8));
        byteArray.splice(0, 8);
        return new UnresolvedMosaicIdDto(unresolvedMosaicId);
    }
    getUnresolvedMosaicId() {
        return this.unresolvedMosaicId;
    }
    getSize() {
        return 8;
    }
    serialize() {
        let newArray = Uint8Array.from([]);
        const unresolvedMosaicIdBytes = GeneratorUtils_1.GeneratorUtils.uint64ToBuffer(this.getUnresolvedMosaicId());
        newArray = GeneratorUtils_1.GeneratorUtils.concatTypedArrays(newArray, unresolvedMosaicIdBytes);
        return newArray;
    }
}
exports.UnresolvedMosaicIdDto = UnresolvedMosaicIdDto;
