// symbol-sdk/dist/src/infrastructure/RepositoryFactoryHttp.js
"use strict";
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
Object.defineProperty(exports, "__esModule", { value: true });
const rxjs_1 = require("rxjs");
const operators_1 = require("rxjs/operators");
const AccountHttp_1 = require("./AccountHttp");
const BlockHttp_1 = require("./BlockHttp");
const ChainHttp_1 = require("./ChainHttp");
const Listener_1 = require("./Listener");
const MetadataHttp_1 = require("./MetadataHttp");
const MosaicHttp_1 = require("./MosaicHttp");
const MultisigHttp_1 = require("./MultisigHttp");
const NamespaceHttp_1 = require("./NamespaceHttp");
const NetworkHttp_1 = require("./NetworkHttp");
const NodeHttp_1 = require("./NodeHttp");
const ReceiptHttp_1 = require("./ReceiptHttp");
const RestrictionAccountHttp_1 = require("./RestrictionAccountHttp");
const RestrictionMosaicHttp_1 = require("./RestrictionMosaicHttp");
const TransactionHttp_1 = require("./TransactionHttp");
const TransactionStatusHttp_1 = require("./TransactionStatusHttp");
/**
 * Receipt http repository.
 *
 */
class RepositoryFactoryHttp {
    /**
     * Constructor
     * @param url the server url.
     * @param configs optional repository factory configs
     */
    constructor(url, configs) {
        this.url = url;
        this.fetchApi = configs === null || configs === void 0 ? void 0 : configs.fetchApi;
        this.networkType = (configs === null || configs === void 0 ? void 0 : configs.networkType) ? rxjs_1.of(configs.networkType)
            : this.createNetworkRepository().getNetworkType().pipe(operators_1.shareReplay(1));
        this.generationHash = (configs === null || configs === void 0 ? void 0 : configs.generationHash) ? rxjs_1.of(configs === null || configs === void 0 ? void 0 : configs.generationHash)
            : this.createNodeRepository()
                .getNodeInfo()
                .pipe(operators_1.map((b) => b.networkGenerationHashSeed))
                .pipe(operators_1.shareReplay(1));
        this.websocketUrl = (configs === null || configs === void 0 ? void 0 : configs.websocketUrl) ? configs === null || configs === void 0 ? void 0 : configs.websocketUrl : `${url.replace(/\/$/, '')}/ws`;
        this.websocketInjected = configs === null || configs === void 0 ? void 0 : configs.websocketInjected;
    }
    createAccountRepository() {
        return new AccountHttp_1.AccountHttp(this.url, this.fetchApi);
    }
    createBlockRepository() {
        return new BlockHttp_1.BlockHttp(this.url, this.fetchApi);
    }
    createChainRepository() {
        return new ChainHttp_1.ChainHttp(this.url, this.fetchApi);
    }
    createMetadataRepository() {
        return new MetadataHttp_1.MetadataHttp(this.url, this.fetchApi);
    }
    createMosaicRepository() {
        return new MosaicHttp_1.MosaicHttp(this.url, this.networkType, this.fetchApi);
    }
    createMultisigRepository() {
        return new MultisigHttp_1.MultisigHttp(this.url, this.fetchApi);
    }
    createNamespaceRepository() {
        return new NamespaceHttp_1.NamespaceHttp(this.url, this.networkType, this.fetchApi);
    }
    createNetworkRepository() {
        return new NetworkHttp_1.NetworkHttp(this.url, this.fetchApi);
    }
    createNodeRepository() {
        return new NodeHttp_1.NodeHttp(this.url, this.fetchApi);
    }
    createReceiptRepository() {
        return new ReceiptHttp_1.ReceiptHttp(this.url, this.fetchApi);
    }
    createRestrictionAccountRepository() {
        return new RestrictionAccountHttp_1.RestrictionAccountHttp(this.url, this.fetchApi);
    }
    createRestrictionMosaicRepository() {
        return new RestrictionMosaicHttp_1.RestrictionMosaicHttp(this.url, this.fetchApi);
    }
    createTransactionRepository() {
        return new TransactionHttp_1.TransactionHttp(this.url, this.fetchApi);
    }
    createTransactionStatusRepository() {
        return new TransactionStatusHttp_1.TransactionStatusHttp(this.url, this.fetchApi);
    }
    getGenerationHash() {
        return this.generationHash;
    }
    getNetworkType() {
        return this.networkType;
    }
    createListener() {
        return new Listener_1.Listener(this.websocketUrl, this.createNamespaceRepository(), this.websocketInjected);
    }
}
exports.RepositoryFactoryHttp = RepositoryFactoryHttp;
