// symbol-openapi-typescript-fetch-client/src/apis/NamespaceRoutesApi.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


import * as runtime from '../runtime';
import {
    AccountsNamesDTO,
    AccountsNamesDTOFromJSON,
    AccountsNamesDTOToJSON,
    Addresses,
    AddressesFromJSON,
    AddressesToJSON,
    AliasTypeEnum,
    AliasTypeEnumFromJSON,
    AliasTypeEnumToJSON,
    ModelError,
    ModelErrorFromJSON,
    ModelErrorToJSON,
    MosaicIds,
    MosaicIdsFromJSON,
    MosaicIdsToJSON,
    MosaicsNamesDTO,
    MosaicsNamesDTOFromJSON,
    MosaicsNamesDTOToJSON,
    NamespaceIds,
    NamespaceIdsFromJSON,
    NamespaceIdsToJSON,
    NamespaceInfoDTO,
    NamespaceInfoDTOFromJSON,
    NamespaceInfoDTOToJSON,
    NamespaceNameDTO,
    NamespaceNameDTOFromJSON,
    NamespaceNameDTOToJSON,
    NamespacePage,
    NamespacePageFromJSON,
    NamespacePageToJSON,
    NamespaceRegistrationTypeEnum,
    NamespaceRegistrationTypeEnumFromJSON,
    NamespaceRegistrationTypeEnumToJSON,
    Order,
    OrderFromJSON,
    OrderToJSON,
} from '../models';

export interface GetAccountsNamesRequest {
    addresses: Addresses;
}

export interface GetMosaicsNamesRequest {
    mosaicIds: MosaicIds;
}

export interface GetNamespaceRequest {
    namespaceId: string;
}

export interface GetNamespacesNamesRequest {
    namespaceIds: NamespaceIds;
}

export interface SearchNamespacesRequest {
    ownerAddress?: string;
    registrationType?: NamespaceRegistrationTypeEnum;
    level0?: string;
    aliasType?: AliasTypeEnum;
    pageSize?: number;
    pageNumber?: number;
    offset?: string;
    order?: Order;
}

/**
 * 
 */
export class NamespaceRoutesApi extends runtime.BaseAPI {

    /**
     * Returns friendly names for accounts.
     * Get readable names for a set of accountIds
     */
    async getAccountsNamesRaw(requestParameters: GetAccountsNamesRequest): Promise<runtime.ApiResponse<AccountsNamesDTO>> {
        if (requestParameters.addresses === null || requestParameters.addresses === undefined) {
            throw new runtime.RequiredError('addresses','Required parameter requestParameters.addresses was null or undefined when calling getAccountsNames.');
        }

        const queryParameters: runtime.HTTPQuery = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        const response = await this.request({
            path: `/namespaces/account/names`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: AddressesToJSON(requestParameters.addresses),
        });

        return new runtime.JSONApiResponse(response, (jsonValue) => AccountsNamesDTOFromJSON(jsonValue));
    }

    /**
     * Returns friendly names for accounts.
     * Get readable names for a set of accountIds
     */
    async getAccountsNames(addresses: Addresses): Promise<AccountsNamesDTO> {
        const response = await this.getAccountsNamesRaw({ addresses: addresses });
        return await response.value();
    }

    /**
     * Returns friendly names for mosaics.
     * Get readable names for a set of mosaics
     */
    async getMosaicsNamesRaw(requestParameters: GetMosaicsNamesRequest): Promise<runtime.ApiResponse<MosaicsNamesDTO>> {
        if (requestParameters.mosaicIds === null || requestParameters.mosaicIds === undefined) {
            throw new runtime.RequiredError('mosaicIds','Required parameter requestParameters.mosaicIds was null or undefined when calling getMosaicsNames.');
        }

        const queryParameters: runtime.HTTPQuery = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        const response = await this.request({
            path: `/namespaces/mosaic/names`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: MosaicIdsToJSON(requestParameters.mosaicIds),
        });

        return new runtime.JSONApiResponse(response, (jsonValue) => MosaicsNamesDTOFromJSON(jsonValue));
    }

    /**
     * Returns friendly names for mosaics.
     * Get readable names for a set of mosaics
     */
    async getMosaicsNames(mosaicIds: MosaicIds): Promise<MosaicsNamesDTO> {
        const response = await this.getMosaicsNamesRaw({ mosaicIds: mosaicIds });
        return await response.value();
    }

    /**
     * Gets the namespace for a given namespace identifier.
     * Get namespace information
     */
    async getNamespaceRaw(requestParameters: GetNamespaceRequest): Promise<runtime.ApiResponse<NamespaceInfoDTO>> {
        if (requestParameters.namespaceId === null || requestParameters.namespaceId === undefined) {
            throw new runtime.RequiredError('namespaceId','Required parameter requestParameters.namespaceId was null or undefined when calling getNamespace.');
        }

        const queryParameters: runtime.HTTPQuery = {};

        const headerParameters: runtime.HTTPHeaders = {};

        const response = await this.request({
            path: `/namespaces/{namespaceId}`.replace(`{${"namespaceId"}}`, encodeURIComponent(String(requestParameters.namespaceId))),
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        });

        return new runtime.JSONApiResponse(response, (jsonValue) => NamespaceInfoDTOFromJSON(jsonValue));
    }

    /**
     * Gets the namespace for a given namespace identifier.
     * Get namespace information
     */
    async getNamespace(namespaceId: string): Promise<NamespaceInfoDTO> {
        const response = await this.getNamespaceRaw({ namespaceId: namespaceId });
        return await response.value();
    }

    /**
     * Returns friendly names for namespaces.
     * Get readable names for a set of namespaces
     */
    async getNamespacesNamesRaw(requestParameters: GetNamespacesNamesRequest): Promise<runtime.ApiResponse<Array<NamespaceNameDTO>>> {
        if (requestParameters.namespaceIds === null || requestParameters.namespaceIds === undefined) {
            throw new runtime.RequiredError('namespaceIds','Required parameter requestParameters.namespaceIds was null or undefined when calling getNamespacesNames.');
        }

        const queryParameters: runtime.HTTPQuery = {};

        const headerParameters: runtime.HTTPHeaders = {};

        headerParameters['Content-Type'] = 'application/json';

        const response = await this.request({
            path: `/namespaces/names`,
            method: 'POST',
            headers: headerParameters,
            query: queryParameters,
            body: NamespaceIdsToJSON(requestParameters.namespaceIds),
        });

        return new runtime.JSONApiResponse(response, (jsonValue) => jsonValue.map(NamespaceNameDTOFromJSON));
    }

    /**
     * Returns friendly names for namespaces.
     * Get readable names for a set of namespaces
     */
    async getNamespacesNames(namespaceIds: NamespaceIds): Promise<Array<NamespaceNameDTO>> {
        const response = await this.getNamespacesNamesRaw({ namespaceIds: namespaceIds });
        return await response.value();
    }

    /**
     * Gets an array of namespaces.
     * Search namespaces
     */
    async searchNamespacesRaw(requestParameters: SearchNamespacesRequest): Promise<runtime.ApiResponse<NamespacePage>> {
        const queryParameters: runtime.HTTPQuery = {};

        if (requestParameters.ownerAddress !== undefined) {
            queryParameters['ownerAddress'] = requestParameters.ownerAddress;
        }

        if (requestParameters.registrationType !== undefined) {
            queryParameters['registrationType'] = requestParameters.registrationType;
        }

        if (requestParameters.level0 !== undefined) {
            queryParameters['level0'] = requestParameters.level0;
        }

        if (requestParameters.aliasType !== undefined) {
            queryParameters['aliasType'] = requestParameters.aliasType;
        }

        if (requestParameters.pageSize !== undefined) {
            queryParameters['pageSize'] = requestParameters.pageSize;
        }

        if (requestParameters.pageNumber !== undefined) {
            queryParameters['pageNumber'] = requestParameters.pageNumber;
        }

        if (requestParameters.offset !== undefined) {
            queryParameters['offset'] = requestParameters.offset;
        }

        if (requestParameters.order !== undefined) {
            queryParameters['order'] = requestParameters.order;
        }

        const headerParameters: runtime.HTTPHeaders = {};

        const response = await this.request({
            path: `/namespaces`,
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        });

        return new runtime.JSONApiResponse(response, (jsonValue) => NamespacePageFromJSON(jsonValue));
    }

    /**
     * Gets an array of namespaces.
     * Search namespaces
     */
    async searchNamespaces(ownerAddress?: string, registrationType?: NamespaceRegistrationTypeEnum, level0?: string, aliasType?: AliasTypeEnum, pageSize?: number, pageNumber?: number, offset?: string, order?: Order): Promise<NamespacePage> {
        const response = await this.searchNamespacesRaw({ ownerAddress: ownerAddress, registrationType: registrationType, level0: level0, aliasType: aliasType, pageSize: pageSize, pageNumber: pageNumber, offset: offset, order: order });
        return await response.value();
    }

}
