// symbol-openapi-typescript-fetch-client/src/apis/SecretLockRoutesApi.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


import * as runtime from '../runtime';
import {
    ModelError,
    ModelErrorFromJSON,
    ModelErrorToJSON,
    Order,
    OrderFromJSON,
    OrderToJSON,
    SecretLockInfoDTO,
    SecretLockInfoDTOFromJSON,
    SecretLockInfoDTOToJSON,
    SecretLockPage,
    SecretLockPageFromJSON,
    SecretLockPageToJSON,
} from '../models';

export interface GetSecretLockRequest {
    secret: string;
}

export interface SearchSecretLockRequest {
    address: string;
    pageSize?: number;
    pageNumber?: number;
    offset?: string;
    order?: Order;
}

/**
 * 
 */
export class SecretLockRoutesApi extends runtime.BaseAPI {

    /**
     * Gets the secret lock for a given secret.
     * Get secret lock information
     */
    async getSecretLockRaw(requestParameters: GetSecretLockRequest): Promise<runtime.ApiResponse<SecretLockInfoDTO>> {
        if (requestParameters.secret === null || requestParameters.secret === undefined) {
            throw new runtime.RequiredError('secret','Required parameter requestParameters.secret was null or undefined when calling getSecretLock.');
        }

        const queryParameters: runtime.HTTPQuery = {};

        const headerParameters: runtime.HTTPHeaders = {};

        const response = await this.request({
            path: `/lock/secret/{secret}`.replace(`{${"secret"}}`, encodeURIComponent(String(requestParameters.secret))),
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        });

        return new runtime.JSONApiResponse(response, (jsonValue) => SecretLockInfoDTOFromJSON(jsonValue));
    }

    /**
     * Gets the secret lock for a given secret.
     * Get secret lock information
     */
    async getSecretLock(secret: string): Promise<SecretLockInfoDTO> {
        const response = await this.getSecretLockRaw({ secret: secret });
        return await response.value();
    }

    /**
     * Returns an array of secret locks.
     * Search secret lock entries
     */
    async searchSecretLockRaw(requestParameters: SearchSecretLockRequest): Promise<runtime.ApiResponse<SecretLockPage>> {
        if (requestParameters.address === null || requestParameters.address === undefined) {
            throw new runtime.RequiredError('address','Required parameter requestParameters.address was null or undefined when calling searchSecretLock.');
        }

        const queryParameters: runtime.HTTPQuery = {};

        if (requestParameters.pageSize !== undefined) {
            queryParameters['pageSize'] = requestParameters.pageSize;
        }

        if (requestParameters.pageNumber !== undefined) {
            queryParameters['pageNumber'] = requestParameters.pageNumber;
        }

        if (requestParameters.offset !== undefined) {
            queryParameters['offset'] = requestParameters.offset;
        }

        if (requestParameters.order !== undefined) {
            queryParameters['order'] = requestParameters.order;
        }

        const headerParameters: runtime.HTTPHeaders = {};

        const response = await this.request({
            path: `/account/{address}/lock/secret`.replace(`{${"address"}}`, encodeURIComponent(String(requestParameters.address))),
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        });

        return new runtime.JSONApiResponse(response, (jsonValue) => SecretLockPageFromJSON(jsonValue));
    }

    /**
     * Returns an array of secret locks.
     * Search secret lock entries
     */
    async searchSecretLock(address: string, pageSize?: number, pageNumber?: number, offset?: string, order?: Order): Promise<SecretLockPage> {
        const response = await this.searchSecretLockRaw({ address: address, pageSize: pageSize, pageNumber: pageNumber, offset: offset, order: order });
        return await response.value();
    }

}
