// @geens_npo/defined-id/lib/handlers/hash/HashRegisterHandler.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.HashRegisterHandler = void 0;
const BaseHandler_1 = require("../BaseHandler");
class HashRegisterHandler extends BaseHandler_1.BaseHandler {
    handleCore(request) {
        let documentHash = request.documentHash;
        let nemCatapultBlockchain = request.nemCatapultBlockchain;
        let nodeUri = request.nodeUri;
        if (documentHash)
            documentHash = documentHash.trim();
        if (nemCatapultBlockchain)
            nemCatapultBlockchain = nemCatapultBlockchain.trim();
        if (nodeUri)
            nodeUri = nodeUri.trim();
        return {
            documentHash: documentHash,
            nemCatapultBlockchain: nemCatapultBlockchain,
            nodeUri: nodeUri
        };
    }
    validateCore(request) {
        const errors = [];
        if (!request.documentHash)
            errors.push("documentHash is required");
        if (!request.nemCatapultBlockchain)
            errors.push("nemCatapultBlockchain is required");
        if (!request.nodeUri)
            errors.push("nodeUri is required");
        return errors;
    }
}
exports.HashRegisterHandler = HashRegisterHandler;
