// symbol-sdk/dist/src/model/node/NodeInfo.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
/**
 * The node info structure describes basic information of a node.
 */
class NodeInfo {
    /**
     * @param publicKey
     * @param networkGenerationHashSeed
     * @param port
     * @param networkIdentifier
     * @param version
     * @param roles
     * @param host
     * @param friendlyName
     */
    constructor(
    /**
     * The public key used to identify the node.
     */
    publicKey, 
    /**
     * The network generation hash seed
     */
    networkGenerationHashSeed, 
    /**
     * The port used for the communication.
     */
    port, 
    /**
     * The network identifier.
     */
    networkIdentifier, 
    /**
     * The version of the application.
     */
    version, 
    /**
     * The roles of the application.
     */
    roles, 
    /**
     * The IP address of the endpoint.
     */
    host, 
    /**
     * The name of the node.
     */
    friendlyName) {
        this.publicKey = publicKey;
        this.networkGenerationHashSeed = networkGenerationHashSeed;
        this.port = port;
        this.networkIdentifier = networkIdentifier;
        this.version = version;
        this.roles = roles;
        this.host = host;
        this.friendlyName = friendlyName;
    }
}
exports.NodeInfo = NodeInfo;
