// symbol-openapi-typescript-fetch-client/src/models/AccountLinkVotingKeyDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
/**
 * 
 * @export
 * @interface AccountLinkVotingKeyDTO
 */
export interface AccountLinkVotingKeyDTO {
    /**
     * 
     * @type {string}
     * @memberof AccountLinkVotingKeyDTO
     */
    publicKey: string;
    /**
     * 
     * @type {string}
     * @memberof AccountLinkVotingKeyDTO
     */
    startPoint: string;
    /**
     * 
     * @type {string}
     * @memberof AccountLinkVotingKeyDTO
     */
    endPoint: string;
}

export function AccountLinkVotingKeyDTOFromJSON(json: any): AccountLinkVotingKeyDTO {
    return AccountLinkVotingKeyDTOFromJSONTyped(json, false);
}

export function AccountLinkVotingKeyDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): AccountLinkVotingKeyDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'publicKey': json['publicKey'],
        'startPoint': json['startPoint'],
        'endPoint': json['endPoint'],
    };
}

export function AccountLinkVotingKeyDTOToJSON(value?: AccountLinkVotingKeyDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'publicKey': value.publicKey,
        'startPoint': value.startPoint,
        'endPoint': value.endPoint,
    };
}
