// symbol-openapi-typescript-fetch-client/src/models/AccountRestrictionDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    AccountRestrictionFlagsEnum,
    AccountRestrictionFlagsEnumFromJSON,
    AccountRestrictionFlagsEnumFromJSONTyped,
    AccountRestrictionFlagsEnumToJSON,
    TransactionTypeEnum,
    TransactionTypeEnumFromJSON,
    TransactionTypeEnumFromJSONTyped,
    TransactionTypeEnumToJSON,
} from './';

/**
 * 
 * @export
 * @interface AccountRestrictionDTO
 */
export interface AccountRestrictionDTO {
    /**
     * 
     * @type {AccountRestrictionFlagsEnum}
     * @memberof AccountRestrictionDTO
     */
    restrictionFlags: AccountRestrictionFlagsEnum;
    /**
     * Address, mosaic id, or transaction type to restrict.
     * @type {Array<string | TransactionTypeEnum>}
     * @memberof AccountRestrictionDTO
     */
    values: Array<string | TransactionTypeEnum>;
}

export function AccountRestrictionDTOFromJSON(json: any): AccountRestrictionDTO {
    return AccountRestrictionDTOFromJSONTyped(json, false);
}

export function AccountRestrictionDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): AccountRestrictionDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'restrictionFlags': json['restrictionFlags'],
        'values': ((json['values'] as Array<any>)),
    };
}

export function AccountRestrictionDTOToJSON(value?: AccountRestrictionDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'restrictionFlags': value.restrictionFlags,
        'values': ((value.values as Array<any>)),
    };
}
