// symbol-openapi-typescript-fetch-client/src/models/BalanceTransferReceiptDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    BalanceTransferReceiptDTOAllOf,
    BalanceTransferReceiptDTOAllOfFromJSON,
    BalanceTransferReceiptDTOAllOfFromJSONTyped,
    BalanceTransferReceiptDTOAllOfToJSON,
    ReceiptDTO,
    ReceiptDTOFromJSON,
    ReceiptDTOFromJSONTyped,
    ReceiptDTOToJSON,
    ReceiptTypeEnum,
    ReceiptTypeEnumFromJSON,
    ReceiptTypeEnumFromJSONTyped,
    ReceiptTypeEnumToJSON,
} from './';

/**
 * Receipt stored when a state change that triggered a mosaic transfer.
 * @export
 * @interface BalanceTransferReceiptDTO
 */
export interface BalanceTransferReceiptDTO {
    /**
     * Version of the receipt.
     * @type {number}
     * @memberof BalanceTransferReceiptDTO
     */
    version: number;
    /**
     * 
     * @type {ReceiptTypeEnum}
     * @memberof BalanceTransferReceiptDTO
     */
    type: ReceiptTypeEnum;
    /**
     * Mosaic identifier.
     * @type {string}
     * @memberof BalanceTransferReceiptDTO
     */
    mosaicId: string;
    /**
     * Absolute amount. An amount of 123456789 (absolute) for a mosaic with divisibility 6 means 123.456789 (relative).
     * @type {string}
     * @memberof BalanceTransferReceiptDTO
     */
    amount: string;
    /**
     * Address expressed in hexadecimal base.
     * @type {string}
     * @memberof BalanceTransferReceiptDTO
     */
    senderAddress: string;
    /**
     * Address expressed in hexadecimal base.
     * @type {string}
     * @memberof BalanceTransferReceiptDTO
     */
    recipientAddress: string;
}

export function BalanceTransferReceiptDTOFromJSON(json: any): BalanceTransferReceiptDTO {
    return BalanceTransferReceiptDTOFromJSONTyped(json, false);
}

export function BalanceTransferReceiptDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): BalanceTransferReceiptDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'version': json['version'],
        'type': json['type'],
        'mosaicId': json['mosaicId'],
        'amount': json['amount'],
        'senderAddress': json['senderAddress'],
        'recipientAddress': json['recipientAddress'],
    };
}

export function BalanceTransferReceiptDTOToJSON(value?: BalanceTransferReceiptDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'version': value.version,
        'type': value.type,
        'mosaicId': value.mosaicId,
        'amount': value.amount,
        'senderAddress': value.senderAddress,
        'recipientAddress': value.recipientAddress,
    };
}
