// symbol-openapi-typescript-fetch-client/src/models/ChainScoreDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
/**
 * 
 * @export
 * @interface ChainScoreDTO
 */
export interface ChainScoreDTO {
    /**
     * Score of the blockchain. During synchronization, nodes try to get the
     * blockchain with highest score in the network.
     * @type {string}
     * @memberof ChainScoreDTO
     */
    scoreHigh: string;
    /**
     * Score of the blockchain. During synchronization, nodes try to get the
     * blockchain with highest score in the network.
     * @type {string}
     * @memberof ChainScoreDTO
     */
    scoreLow: string;
}

export function ChainScoreDTOFromJSON(json: any): ChainScoreDTO {
    return ChainScoreDTOFromJSONTyped(json, false);
}

export function ChainScoreDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): ChainScoreDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'scoreHigh': json['scoreHigh'],
        'scoreLow': json['scoreLow'],
    };
}

export function ChainScoreDTOToJSON(value?: ChainScoreDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'scoreHigh': value.scoreHigh,
        'scoreLow': value.scoreLow,
    };
}
