// symbol-openapi-typescript-fetch-client/src/models/EmbeddedTransferTransactionDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    EmbeddedTransactionDTO,
    EmbeddedTransactionDTOFromJSON,
    EmbeddedTransactionDTOFromJSONTyped,
    EmbeddedTransactionDTOToJSON,
    MessageDTO,
    MessageDTOFromJSON,
    MessageDTOFromJSONTyped,
    MessageDTOToJSON,
    NetworkTypeEnum,
    NetworkTypeEnumFromJSON,
    NetworkTypeEnumFromJSONTyped,
    NetworkTypeEnumToJSON,
    TransferTransactionBodyDTO,
    TransferTransactionBodyDTOFromJSON,
    TransferTransactionBodyDTOFromJSONTyped,
    TransferTransactionBodyDTOToJSON,
    UnresolvedMosaic,
    UnresolvedMosaicFromJSON,
    UnresolvedMosaicFromJSONTyped,
    UnresolvedMosaicToJSON,
} from './';

/**
 * 
 * @export
 * @interface EmbeddedTransferTransactionDTO
 */
export interface EmbeddedTransferTransactionDTO {
    /**
     * Public key.
     * @type {string}
     * @memberof EmbeddedTransferTransactionDTO
     */
    signerPublicKey: string;
    /**
     * Entity version.
     * @type {number}
     * @memberof EmbeddedTransferTransactionDTO
     */
    version: number;
    /**
     * 
     * @type {NetworkTypeEnum}
     * @memberof EmbeddedTransferTransactionDTO
     */
    network: NetworkTypeEnum;
    /**
     * 
     * @type {number}
     * @memberof EmbeddedTransferTransactionDTO
     */
    type: number;
    /**
     * Address expressed in hexadecimal base. If the bit 0 of byte 0 is not set (like in 0x90), then it is a
     * regular address. Else (e.g. 0x91) it represents a namespace id which starts at byte 1.
     * @type {string}
     * @memberof EmbeddedTransferTransactionDTO
     */
    recipientAddress: string;
    /**
     * Array of mosaics sent to the recipient.
     * @type {Array<UnresolvedMosaic>}
     * @memberof EmbeddedTransferTransactionDTO
     */
    mosaics: Array<UnresolvedMosaic>;
    /**
     * 
     * @type {MessageDTO}
     * @memberof EmbeddedTransferTransactionDTO
     */
    message: MessageDTO;
}

export function EmbeddedTransferTransactionDTOFromJSON(json: any): EmbeddedTransferTransactionDTO {
    return EmbeddedTransferTransactionDTOFromJSONTyped(json, false);
}

export function EmbeddedTransferTransactionDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): EmbeddedTransferTransactionDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'signerPublicKey': json['signerPublicKey'],
        'version': json['version'],
        'network': json['network'],
        'type': json['type'],
        'recipientAddress': json['recipientAddress'],
        'mosaics': ((json['mosaics'] as Array<any>)),
        'message': json['message'],
    };
}

export function EmbeddedTransferTransactionDTOToJSON(value?: EmbeddedTransferTransactionDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'signerPublicKey': value.signerPublicKey,
        'version': value.version,
        'network': value.network,
        'type': value.type,
        'recipientAddress': value.recipientAddress,
        'mosaics': ((value.mosaics as Array<any>)),
        'message': value.message,
    };
}
