// symbol-openapi-typescript-fetch-client/src/models/MessageDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    MessageTypeEnum,
    MessageTypeEnumFromJSON,
    MessageTypeEnumFromJSONTyped,
    MessageTypeEnumToJSON,
} from './';

/**
 * 
 * @export
 * @interface MessageDTO
 */
export interface MessageDTO {
    /**
     * 
     * @type {MessageTypeEnum}
     * @memberof MessageDTO
     */
    type: MessageTypeEnum;
    /**
     * Content of the message.
     * @type {string}
     * @memberof MessageDTO
     */
    payload: string;
}

export function MessageDTOFromJSON(json: any): MessageDTO {
    return MessageDTOFromJSONTyped(json, false);
}

export function MessageDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): MessageDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'type': json['type'],
        'payload': json['payload'],
    };
}

export function MessageDTOToJSON(value?: MessageDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'type': value.type,
        'payload': value.payload,
    };
}
