// symbol-openapi-typescript-fetch-client/src/models/MosaicGlobalRestrictionEntryWrapperDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    MosaicGlobalRestrictionEntryDTO,
    MosaicGlobalRestrictionEntryDTOFromJSON,
    MosaicGlobalRestrictionEntryDTOFromJSONTyped,
    MosaicGlobalRestrictionEntryDTOToJSON,
    MosaicRestrictionEntryTypeEnum,
    MosaicRestrictionEntryTypeEnumFromJSON,
    MosaicRestrictionEntryTypeEnumFromJSONTyped,
    MosaicRestrictionEntryTypeEnumToJSON,
} from './';

/**
 * 
 * @export
 * @interface MosaicGlobalRestrictionEntryWrapperDTO
 */
export interface MosaicGlobalRestrictionEntryWrapperDTO {
    /**
     * 
     * @type {string}
     * @memberof MosaicGlobalRestrictionEntryWrapperDTO
     */
    compositeHash: string;
    /**
     * 
     * @type {MosaicRestrictionEntryTypeEnum}
     * @memberof MosaicGlobalRestrictionEntryWrapperDTO
     */
    entryType: MosaicRestrictionEntryTypeEnum;
    /**
     * Mosaic identifier.
     * @type {string}
     * @memberof MosaicGlobalRestrictionEntryWrapperDTO
     */
    mosaicId: string;
    /**
     * 
     * @type {Array<MosaicGlobalRestrictionEntryDTO>}
     * @memberof MosaicGlobalRestrictionEntryWrapperDTO
     */
    restrictions: Array<MosaicGlobalRestrictionEntryDTO>;
}

export function MosaicGlobalRestrictionEntryWrapperDTOFromJSON(json: any): MosaicGlobalRestrictionEntryWrapperDTO {
    return MosaicGlobalRestrictionEntryWrapperDTOFromJSONTyped(json, false);
}

export function MosaicGlobalRestrictionEntryWrapperDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): MosaicGlobalRestrictionEntryWrapperDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'compositeHash': json['compositeHash'],
        'entryType': json['entryType'],
        'mosaicId': json['mosaicId'],
        'restrictions': ((json['restrictions'] as Array<any>)),
    };
}

export function MosaicGlobalRestrictionEntryWrapperDTOToJSON(value?: MosaicGlobalRestrictionEntryWrapperDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'compositeHash': value.compositeHash,
        'entryType': value.entryType,
        'mosaicId': value.mosaicId,
        'restrictions': ((value.restrictions as Array<any>)),
    };
}
