// symbol-openapi-typescript-fetch-client/src/models/NamespaceExpiryReceiptDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    NamespaceExpiryReceiptDTOAllOf,
    NamespaceExpiryReceiptDTOAllOfFromJSON,
    NamespaceExpiryReceiptDTOAllOfFromJSONTyped,
    NamespaceExpiryReceiptDTOAllOfToJSON,
    ReceiptDTO,
    ReceiptDTOFromJSON,
    ReceiptDTOFromJSONTyped,
    ReceiptDTOToJSON,
    ReceiptTypeEnum,
    ReceiptTypeEnumFromJSON,
    ReceiptTypeEnumFromJSONTyped,
    ReceiptTypeEnumToJSON,
} from './';

/**
 * Receipt stored when a namespace expires.
 * @export
 * @interface NamespaceExpiryReceiptDTO
 */
export interface NamespaceExpiryReceiptDTO {
    /**
     * Version of the receipt.
     * @type {number}
     * @memberof NamespaceExpiryReceiptDTO
     */
    version: number;
    /**
     * 
     * @type {ReceiptTypeEnum}
     * @memberof NamespaceExpiryReceiptDTO
     */
    type: ReceiptTypeEnum;
    /**
     * Namespace identifier.
     * @type {string}
     * @memberof NamespaceExpiryReceiptDTO
     */
    artifactId: string;
}

export function NamespaceExpiryReceiptDTOFromJSON(json: any): NamespaceExpiryReceiptDTO {
    return NamespaceExpiryReceiptDTOFromJSONTyped(json, false);
}

export function NamespaceExpiryReceiptDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): NamespaceExpiryReceiptDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'version': json['version'],
        'type': json['type'],
        'artifactId': json['artifactId'],
    };
}

export function NamespaceExpiryReceiptDTOToJSON(value?: NamespaceExpiryReceiptDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'version': value.version,
        'type': value.type,
        'artifactId': value.artifactId,
    };
}
