// symbol-openapi-typescript-fetch-client/src/models/NetworkPropertiesDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    NodeIdentityEqualityStrategy,
    NodeIdentityEqualityStrategyFromJSON,
    NodeIdentityEqualityStrategyFromJSONTyped,
    NodeIdentityEqualityStrategyToJSON,
} from './';

/**
 * Network related configuration properties.
 * @export
 * @interface NetworkPropertiesDTO
 */
export interface NetworkPropertiesDTO {
    /**
     * Network identifier.
     * @type {string}
     * @memberof NetworkPropertiesDTO
     */
    identifier?: string;
    /**
     * 
     * @type {NodeIdentityEqualityStrategy}
     * @memberof NetworkPropertiesDTO
     */
    nodeEqualityStrategy?: NodeIdentityEqualityStrategy;
    /**
     * Public key.
     * @type {string}
     * @memberof NetworkPropertiesDTO
     */
    nemesisSignerPublicKey?: string;
    /**
     * 
     * @type {string}
     * @memberof NetworkPropertiesDTO
     */
    generationHashSeed?: string;
    /**
     * Nemesis epoch time adjustment.
     * @type {string}
     * @memberof NetworkPropertiesDTO
     */
    epochAdjustment?: string;
}

export function NetworkPropertiesDTOFromJSON(json: any): NetworkPropertiesDTO {
    return NetworkPropertiesDTOFromJSONTyped(json, false);
}

export function NetworkPropertiesDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): NetworkPropertiesDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'identifier': !exists(json, 'identifier') ? undefined : json['identifier'],
        'nodeEqualityStrategy': !exists(json, 'nodeEqualityStrategy') ? undefined : json['nodeEqualityStrategy'],
        'nemesisSignerPublicKey': !exists(json, 'nemesisSignerPublicKey') ? undefined : json['nemesisSignerPublicKey'],
        'generationHashSeed': !exists(json, 'generationHashSeed') ? undefined : json['generationHashSeed'],
        'epochAdjustment': !exists(json, 'epochAdjustment') ? undefined : json['epochAdjustment'],
    };
}

export function NetworkPropertiesDTOToJSON(value?: NetworkPropertiesDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'identifier': value.identifier,
        'nodeEqualityStrategy': value.nodeEqualityStrategy,
        'nemesisSignerPublicKey': value.nemesisSignerPublicKey,
        'generationHashSeed': value.generationHashSeed,
        'epochAdjustment': value.epochAdjustment,
    };
}
