// symbol-openapi-typescript-fetch-client/src/models/NetworkTypeEnum.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

/**
 * Network type:
 * * 0x60 (96 decimal) - Private network.
 * * 0x90 (144 decimal) - Private test network.
 * * 0x68 (104 decimal) - Public main network.
 * * 0x98 (152 decimal) - Public test network.
 * @export
 * @enum {string}
 */
export enum NetworkTypeEnum {
    NUMBER_104 = 104,
    NUMBER_152 = 152,
    NUMBER_96 = 96,
    NUMBER_144 = 144
}

export function NetworkTypeEnumFromJSON(json: any): NetworkTypeEnum {
    return NetworkTypeEnumFromJSONTyped(json, false);
}

export function NetworkTypeEnumFromJSONTyped(json: any, ignoreDiscriminator: boolean): NetworkTypeEnum {
    return json as NetworkTypeEnum;
}

export function NetworkTypeEnumToJSON(value?: NetworkTypeEnum | null): any {
    return value as any;
}
