// symbol-openapi-typescript-fetch-client/src/models/NodeIdentityEqualityStrategy.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

/**
 * Node equality strategy. Defines if the identifier for the node must be its public key or host.
 * @export
 * @enum {string}
 */
export enum NodeIdentityEqualityStrategy {
    Host = 'host',
    PublicKey = 'public-key'
}

export function NodeIdentityEqualityStrategyFromJSON(json: any): NodeIdentityEqualityStrategy {
    return NodeIdentityEqualityStrategyFromJSONTyped(json, false);
}

export function NodeIdentityEqualityStrategyFromJSONTyped(json: any, ignoreDiscriminator: boolean): NodeIdentityEqualityStrategy {
    return json as NodeIdentityEqualityStrategy;
}

export function NodeIdentityEqualityStrategyToJSON(value?: NodeIdentityEqualityStrategy | null): any {
    return value as any;
}
