// symbol-openapi-typescript-fetch-client/src/models/SecretLockNetworkPropertiesDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
/**
 * 
 * @export
 * @interface SecretLockNetworkPropertiesDTO
 */
export interface SecretLockNetworkPropertiesDTO {
    /**
     * Maximum number of blocks for which a secret lock can exist.
     * @type {string}
     * @memberof SecretLockNetworkPropertiesDTO
     */
    maxSecretLockDuration?: string;
    /**
     * Minimum size of a proof in bytes.
     * @type {string}
     * @memberof SecretLockNetworkPropertiesDTO
     */
    minProofSize?: string;
    /**
     * Maximum size of a proof in bytes.
     * @type {string}
     * @memberof SecretLockNetworkPropertiesDTO
     */
    maxProofSize?: string;
}

export function SecretLockNetworkPropertiesDTOFromJSON(json: any): SecretLockNetworkPropertiesDTO {
    return SecretLockNetworkPropertiesDTOFromJSONTyped(json, false);
}

export function SecretLockNetworkPropertiesDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): SecretLockNetworkPropertiesDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'maxSecretLockDuration': !exists(json, 'maxSecretLockDuration') ? undefined : json['maxSecretLockDuration'],
        'minProofSize': !exists(json, 'minProofSize') ? undefined : json['minProofSize'],
        'maxProofSize': !exists(json, 'maxProofSize') ? undefined : json['maxProofSize'],
    };
}

export function SecretLockNetworkPropertiesDTOToJSON(value?: SecretLockNetworkPropertiesDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'maxSecretLockDuration': value.maxSecretLockDuration,
        'minProofSize': value.minProofSize,
        'maxProofSize': value.maxProofSize,
    };
}
