// symbol-openapi-typescript-fetch-client/src/models/TransferTransactionBodyDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    MessageDTO,
    MessageDTOFromJSON,
    MessageDTOFromJSONTyped,
    MessageDTOToJSON,
    UnresolvedMosaic,
    UnresolvedMosaicFromJSON,
    UnresolvedMosaicFromJSONTyped,
    UnresolvedMosaicToJSON,
} from './';

/**
 * 
 * @export
 * @interface TransferTransactionBodyDTO
 */
export interface TransferTransactionBodyDTO {
    /**
     * Address expressed in hexadecimal base. If the bit 0 of byte 0 is not set (like in 0x90), then it is a
     * regular address. Else (e.g. 0x91) it represents a namespace id which starts at byte 1.
     * @type {string}
     * @memberof TransferTransactionBodyDTO
     */
    recipientAddress: string;
    /**
     * Array of mosaics sent to the recipient.
     * @type {Array<UnresolvedMosaic>}
     * @memberof TransferTransactionBodyDTO
     */
    mosaics: Array<UnresolvedMosaic>;
    /**
     * 
     * @type {MessageDTO}
     * @memberof TransferTransactionBodyDTO
     */
    message: MessageDTO;
}

export function TransferTransactionBodyDTOFromJSON(json: any): TransferTransactionBodyDTO {
    return TransferTransactionBodyDTOFromJSONTyped(json, false);
}

export function TransferTransactionBodyDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): TransferTransactionBodyDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'recipientAddress': json['recipientAddress'],
        'mosaics': ((json['mosaics'] as Array<any>)),
        'message': json['message'],
    };
}

export function TransferTransactionBodyDTOToJSON(value?: TransferTransactionBodyDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'recipientAddress': value.recipientAddress,
        'mosaics': ((value.mosaics as Array<any>)),
        'message': value.message,
    };
}
