// symbol-openapi-typescript-fetch-client/src/models/TransactionStatementDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    BalanceChangeReceiptDTO,
    BalanceChangeReceiptDTOFromJSON,
    BalanceChangeReceiptDTOFromJSONTyped,
    BalanceChangeReceiptDTOToJSON,
    BalanceTransferReceiptDTO,
    BalanceTransferReceiptDTOFromJSON,
    BalanceTransferReceiptDTOFromJSONTyped,
    BalanceTransferReceiptDTOToJSON,
    InflationReceiptDTO,
    InflationReceiptDTOFromJSON,
    InflationReceiptDTOFromJSONTyped,
    InflationReceiptDTOToJSON,
    MosaicExpiryReceiptDTO,
    MosaicExpiryReceiptDTOFromJSON,
    MosaicExpiryReceiptDTOFromJSONTyped,
    MosaicExpiryReceiptDTOToJSON,
    NamespaceExpiryReceiptDTO,
    NamespaceExpiryReceiptDTOFromJSON,
    NamespaceExpiryReceiptDTOFromJSONTyped,
    NamespaceExpiryReceiptDTOToJSON,
    SourceDTO,
    SourceDTOFromJSON,
    SourceDTOFromJSONTyped,
    SourceDTOToJSON,
} from './';

/**
 * 
 * @export
 * @interface TransactionStatementDTO
 */
export interface TransactionStatementDTO {
    /**
     * Height of the blockchain.
     * @type {string}
     * @memberof TransactionStatementDTO
     */
    height: string;
    /**
     * 
     * @type {SourceDTO}
     * @memberof TransactionStatementDTO
     */
    source: SourceDTO;
    /**
     * Array of receipts.
     * @type {Array<BalanceTransferReceiptDTO | BalanceChangeReceiptDTO | NamespaceExpiryReceiptDTO | MosaicExpiryReceiptDTO | InflationReceiptDTO>}
     * @memberof TransactionStatementDTO
     */
    receipts: Array<BalanceTransferReceiptDTO | BalanceChangeReceiptDTO | NamespaceExpiryReceiptDTO | MosaicExpiryReceiptDTO | InflationReceiptDTO>;
}

export function TransactionStatementDTOFromJSON(json: any): TransactionStatementDTO {
    return TransactionStatementDTOFromJSONTyped(json, false);
}

export function TransactionStatementDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): TransactionStatementDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'height': json['height'],
        'source': json['source'],
        'receipts': ((json['receipts'] as Array<any>)),
    };
}

export function TransactionStatementDTOToJSON(value?: TransactionStatementDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'height': value.height,
        'source': value.source,
        'receipts': ((value.receipts as Array<any>)),
    };
}
