// symbol-openapi-typescript-fetch-client/src/models/AccountRestrictionsDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    AccountRestrictionDTO,
    AccountRestrictionDTOFromJSON,
    AccountRestrictionDTOFromJSONTyped,
    AccountRestrictionDTOToJSON,
} from './';

/**
 * 
 * @export
 * @interface AccountRestrictionsDTO
 */
export interface AccountRestrictionsDTO {
    /**
     * Address expressed in hexadecimal base.
     * @type {string}
     * @memberof AccountRestrictionsDTO
     */
    address: string;
    /**
     * 
     * @type {Array<AccountRestrictionDTO>}
     * @memberof AccountRestrictionsDTO
     */
    restrictions: Array<AccountRestrictionDTO>;
}

export function AccountRestrictionsDTOFromJSON(json: any): AccountRestrictionsDTO {
    return AccountRestrictionsDTOFromJSONTyped(json, false);
}

export function AccountRestrictionsDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): AccountRestrictionsDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'address': json['address'],
        'restrictions': ((json['restrictions'] as Array<any>)),
    };
}

export function AccountRestrictionsDTOToJSON(value?: AccountRestrictionsDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'address': value.address,
        'restrictions': ((value.restrictions as Array<any>)),
    };
}
