// symbol-openapi-typescript-fetch-client/src/models/EmbeddedAccountOperationRestrictionTransactionDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    AccountOperationRestrictionTransactionBodyDTO,
    AccountOperationRestrictionTransactionBodyDTOFromJSON,
    AccountOperationRestrictionTransactionBodyDTOFromJSONTyped,
    AccountOperationRestrictionTransactionBodyDTOToJSON,
    AccountRestrictionFlagsEnum,
    AccountRestrictionFlagsEnumFromJSON,
    AccountRestrictionFlagsEnumFromJSONTyped,
    AccountRestrictionFlagsEnumToJSON,
    EmbeddedTransactionDTO,
    EmbeddedTransactionDTOFromJSON,
    EmbeddedTransactionDTOFromJSONTyped,
    EmbeddedTransactionDTOToJSON,
    NetworkTypeEnum,
    NetworkTypeEnumFromJSON,
    NetworkTypeEnumFromJSONTyped,
    NetworkTypeEnumToJSON,
    TransactionTypeEnum,
    TransactionTypeEnumFromJSON,
    TransactionTypeEnumFromJSONTyped,
    TransactionTypeEnumToJSON,
} from './';

/**
 * 
 * @export
 * @interface EmbeddedAccountOperationRestrictionTransactionDTO
 */
export interface EmbeddedAccountOperationRestrictionTransactionDTO {
    /**
     * Public key.
     * @type {string}
     * @memberof EmbeddedAccountOperationRestrictionTransactionDTO
     */
    signerPublicKey: string;
    /**
     * Entity version.
     * @type {number}
     * @memberof EmbeddedAccountOperationRestrictionTransactionDTO
     */
    version: number;
    /**
     * 
     * @type {NetworkTypeEnum}
     * @memberof EmbeddedAccountOperationRestrictionTransactionDTO
     */
    network: NetworkTypeEnum;
    /**
     * 
     * @type {number}
     * @memberof EmbeddedAccountOperationRestrictionTransactionDTO
     */
    type: number;
    /**
     * 
     * @type {AccountRestrictionFlagsEnum}
     * @memberof EmbeddedAccountOperationRestrictionTransactionDTO
     */
    restrictionFlags: AccountRestrictionFlagsEnum;
    /**
     * Account restriction additions.
     * @type {Array<TransactionTypeEnum>}
     * @memberof EmbeddedAccountOperationRestrictionTransactionDTO
     */
    restrictionAdditions: Array<TransactionTypeEnum>;
    /**
     * Account restriction deletions.
     * @type {Array<TransactionTypeEnum>}
     * @memberof EmbeddedAccountOperationRestrictionTransactionDTO
     */
    restrictionDeletions: Array<TransactionTypeEnum>;
}

export function EmbeddedAccountOperationRestrictionTransactionDTOFromJSON(json: any): EmbeddedAccountOperationRestrictionTransactionDTO {
    return EmbeddedAccountOperationRestrictionTransactionDTOFromJSONTyped(json, false);
}

export function EmbeddedAccountOperationRestrictionTransactionDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): EmbeddedAccountOperationRestrictionTransactionDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'signerPublicKey': json['signerPublicKey'],
        'version': json['version'],
        'network': json['network'],
        'type': json['type'],
        'restrictionFlags': json['restrictionFlags'],
        'restrictionAdditions': ((json['restrictionAdditions'] as Array<any>)),
        'restrictionDeletions': ((json['restrictionDeletions'] as Array<any>)),
    };
}

export function EmbeddedAccountOperationRestrictionTransactionDTOToJSON(value?: EmbeddedAccountOperationRestrictionTransactionDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'signerPublicKey': value.signerPublicKey,
        'version': value.version,
        'network': value.network,
        'type': value.type,
        'restrictionFlags': value.restrictionFlags,
        'restrictionAdditions': ((value.restrictionAdditions as Array<any>)),
        'restrictionDeletions': ((value.restrictionDeletions as Array<any>)),
    };
}
