// symbol-openapi-typescript-fetch-client/src/models/EmbeddedMosaicAliasTransactionDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    AliasActionEnum,
    AliasActionEnumFromJSON,
    AliasActionEnumFromJSONTyped,
    AliasActionEnumToJSON,
    EmbeddedTransactionDTO,
    EmbeddedTransactionDTOFromJSON,
    EmbeddedTransactionDTOFromJSONTyped,
    EmbeddedTransactionDTOToJSON,
    MosaicAliasTransactionBodyDTO,
    MosaicAliasTransactionBodyDTOFromJSON,
    MosaicAliasTransactionBodyDTOFromJSONTyped,
    MosaicAliasTransactionBodyDTOToJSON,
    NetworkTypeEnum,
    NetworkTypeEnumFromJSON,
    NetworkTypeEnumFromJSONTyped,
    NetworkTypeEnumToJSON,
} from './';

/**
 * 
 * @export
 * @interface EmbeddedMosaicAliasTransactionDTO
 */
export interface EmbeddedMosaicAliasTransactionDTO {
    /**
     * Public key.
     * @type {string}
     * @memberof EmbeddedMosaicAliasTransactionDTO
     */
    signerPublicKey: string;
    /**
     * Entity version.
     * @type {number}
     * @memberof EmbeddedMosaicAliasTransactionDTO
     */
    version: number;
    /**
     * 
     * @type {NetworkTypeEnum}
     * @memberof EmbeddedMosaicAliasTransactionDTO
     */
    network: NetworkTypeEnum;
    /**
     * 
     * @type {number}
     * @memberof EmbeddedMosaicAliasTransactionDTO
     */
    type: number;
    /**
     * Namespace identifier.
     * @type {string}
     * @memberof EmbeddedMosaicAliasTransactionDTO
     */
    namespaceId: string;
    /**
     * Mosaic identifier.
     * @type {string}
     * @memberof EmbeddedMosaicAliasTransactionDTO
     */
    mosaicId: string;
    /**
     * 
     * @type {AliasActionEnum}
     * @memberof EmbeddedMosaicAliasTransactionDTO
     */
    aliasAction: AliasActionEnum;
}

export function EmbeddedMosaicAliasTransactionDTOFromJSON(json: any): EmbeddedMosaicAliasTransactionDTO {
    return EmbeddedMosaicAliasTransactionDTOFromJSONTyped(json, false);
}

export function EmbeddedMosaicAliasTransactionDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): EmbeddedMosaicAliasTransactionDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'signerPublicKey': json['signerPublicKey'],
        'version': json['version'],
        'network': json['network'],
        'type': json['type'],
        'namespaceId': json['namespaceId'],
        'mosaicId': json['mosaicId'],
        'aliasAction': json['aliasAction'],
    };
}

export function EmbeddedMosaicAliasTransactionDTOToJSON(value?: EmbeddedMosaicAliasTransactionDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'signerPublicKey': value.signerPublicKey,
        'version': value.version,
        'network': value.network,
        'type': value.type,
        'namespaceId': value.namespaceId,
        'mosaicId': value.mosaicId,
        'aliasAction': value.aliasAction,
    };
}
