// symbol-openapi-typescript-fetch-client/src/models/EmbeddedTransactionInfoDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    EmbeddedAccountAddressRestrictionTransactionDTO,
    EmbeddedAccountAddressRestrictionTransactionDTOFromJSON,
    EmbeddedAccountAddressRestrictionTransactionDTOFromJSONTyped,
    EmbeddedAccountAddressRestrictionTransactionDTOToJSON,
    EmbeddedAccountKeyLinkTransactionDTO,
    EmbeddedAccountKeyLinkTransactionDTOFromJSON,
    EmbeddedAccountKeyLinkTransactionDTOFromJSONTyped,
    EmbeddedAccountKeyLinkTransactionDTOToJSON,
    EmbeddedAccountMetadataTransactionDTO,
    EmbeddedAccountMetadataTransactionDTOFromJSON,
    EmbeddedAccountMetadataTransactionDTOFromJSONTyped,
    EmbeddedAccountMetadataTransactionDTOToJSON,
    EmbeddedAccountMosaicRestrictionTransactionDTO,
    EmbeddedAccountMosaicRestrictionTransactionDTOFromJSON,
    EmbeddedAccountMosaicRestrictionTransactionDTOFromJSONTyped,
    EmbeddedAccountMosaicRestrictionTransactionDTOToJSON,
    EmbeddedAccountOperationRestrictionTransactionDTO,
    EmbeddedAccountOperationRestrictionTransactionDTOFromJSON,
    EmbeddedAccountOperationRestrictionTransactionDTOFromJSONTyped,
    EmbeddedAccountOperationRestrictionTransactionDTOToJSON,
    EmbeddedAddressAliasTransactionDTO,
    EmbeddedAddressAliasTransactionDTOFromJSON,
    EmbeddedAddressAliasTransactionDTOFromJSONTyped,
    EmbeddedAddressAliasTransactionDTOToJSON,
    EmbeddedHashLockTransactionDTO,
    EmbeddedHashLockTransactionDTOFromJSON,
    EmbeddedHashLockTransactionDTOFromJSONTyped,
    EmbeddedHashLockTransactionDTOToJSON,
    EmbeddedMosaicAddressRestrictionTransactionDTO,
    EmbeddedMosaicAddressRestrictionTransactionDTOFromJSON,
    EmbeddedMosaicAddressRestrictionTransactionDTOFromJSONTyped,
    EmbeddedMosaicAddressRestrictionTransactionDTOToJSON,
    EmbeddedMosaicAliasTransactionDTO,
    EmbeddedMosaicAliasTransactionDTOFromJSON,
    EmbeddedMosaicAliasTransactionDTOFromJSONTyped,
    EmbeddedMosaicAliasTransactionDTOToJSON,
    EmbeddedMosaicDefinitionTransactionDTO,
    EmbeddedMosaicDefinitionTransactionDTOFromJSON,
    EmbeddedMosaicDefinitionTransactionDTOFromJSONTyped,
    EmbeddedMosaicDefinitionTransactionDTOToJSON,
    EmbeddedMosaicGlobalRestrictionTransactionDTO,
    EmbeddedMosaicGlobalRestrictionTransactionDTOFromJSON,
    EmbeddedMosaicGlobalRestrictionTransactionDTOFromJSONTyped,
    EmbeddedMosaicGlobalRestrictionTransactionDTOToJSON,
    EmbeddedMosaicMetadataTransactionDTO,
    EmbeddedMosaicMetadataTransactionDTOFromJSON,
    EmbeddedMosaicMetadataTransactionDTOFromJSONTyped,
    EmbeddedMosaicMetadataTransactionDTOToJSON,
    EmbeddedMosaicSupplyChangeTransactionDTO,
    EmbeddedMosaicSupplyChangeTransactionDTOFromJSON,
    EmbeddedMosaicSupplyChangeTransactionDTOFromJSONTyped,
    EmbeddedMosaicSupplyChangeTransactionDTOToJSON,
    EmbeddedMultisigAccountModificationTransactionDTO,
    EmbeddedMultisigAccountModificationTransactionDTOFromJSON,
    EmbeddedMultisigAccountModificationTransactionDTOFromJSONTyped,
    EmbeddedMultisigAccountModificationTransactionDTOToJSON,
    EmbeddedNamespaceMetadataTransactionDTO,
    EmbeddedNamespaceMetadataTransactionDTOFromJSON,
    EmbeddedNamespaceMetadataTransactionDTOFromJSONTyped,
    EmbeddedNamespaceMetadataTransactionDTOToJSON,
    EmbeddedNamespaceRegistrationTransactionDTO,
    EmbeddedNamespaceRegistrationTransactionDTOFromJSON,
    EmbeddedNamespaceRegistrationTransactionDTOFromJSONTyped,
    EmbeddedNamespaceRegistrationTransactionDTOToJSON,
    EmbeddedNodeKeyLinkTransactionDTO,
    EmbeddedNodeKeyLinkTransactionDTOFromJSON,
    EmbeddedNodeKeyLinkTransactionDTOFromJSONTyped,
    EmbeddedNodeKeyLinkTransactionDTOToJSON,
    EmbeddedSecretLockTransactionDTO,
    EmbeddedSecretLockTransactionDTOFromJSON,
    EmbeddedSecretLockTransactionDTOFromJSONTyped,
    EmbeddedSecretLockTransactionDTOToJSON,
    EmbeddedSecretProofTransactionDTO,
    EmbeddedSecretProofTransactionDTOFromJSON,
    EmbeddedSecretProofTransactionDTOFromJSONTyped,
    EmbeddedSecretProofTransactionDTOToJSON,
    EmbeddedTransactionMetaDTO,
    EmbeddedTransactionMetaDTOFromJSON,
    EmbeddedTransactionMetaDTOFromJSONTyped,
    EmbeddedTransactionMetaDTOToJSON,
    EmbeddedTransferTransactionDTO,
    EmbeddedTransferTransactionDTOFromJSON,
    EmbeddedTransferTransactionDTOFromJSONTyped,
    EmbeddedTransferTransactionDTOToJSON,
    EmbeddedVotingKeyLinkTransactionDTO,
    EmbeddedVotingKeyLinkTransactionDTOFromJSON,
    EmbeddedVotingKeyLinkTransactionDTOFromJSONTyped,
    EmbeddedVotingKeyLinkTransactionDTOToJSON,
    EmbeddedVrfKeyLinkTransactionDTO,
    EmbeddedVrfKeyLinkTransactionDTOFromJSON,
    EmbeddedVrfKeyLinkTransactionDTOFromJSONTyped,
    EmbeddedVrfKeyLinkTransactionDTOToJSON,
} from './';

/**
 * 
 * @export
 * @interface EmbeddedTransactionInfoDTO
 */
export interface EmbeddedTransactionInfoDTO {
    /**
     * Internal resource identifier.
     * @type {string}
     * @memberof EmbeddedTransactionInfoDTO
     */
    id: string;
    /**
     * 
     * @type {EmbeddedTransactionMetaDTO}
     * @memberof EmbeddedTransactionInfoDTO
     */
    meta: EmbeddedTransactionMetaDTO;
    /**
     * 
     * @type {EmbeddedAccountKeyLinkTransactionDTO | EmbeddedNodeKeyLinkTransactionDTO | EmbeddedVrfKeyLinkTransactionDTO | EmbeddedVotingKeyLinkTransactionDTO | EmbeddedHashLockTransactionDTO | EmbeddedSecretLockTransactionDTO | EmbeddedSecretProofTransactionDTO | EmbeddedAccountMetadataTransactionDTO | EmbeddedMosaicMetadataTransactionDTO | EmbeddedNamespaceMetadataTransactionDTO | EmbeddedMosaicDefinitionTransactionDTO | EmbeddedMosaicSupplyChangeTransactionDTO | EmbeddedNamespaceRegistrationTransactionDTO | EmbeddedAddressAliasTransactionDTO | EmbeddedMosaicAliasTransactionDTO | EmbeddedMultisigAccountModificationTransactionDTO | EmbeddedAccountAddressRestrictionTransactionDTO | EmbeddedAccountMosaicRestrictionTransactionDTO | EmbeddedAccountOperationRestrictionTransactionDTO | EmbeddedMosaicGlobalRestrictionTransactionDTO | EmbeddedMosaicAddressRestrictionTransactionDTO | EmbeddedTransferTransactionDTO}
     * @memberof EmbeddedTransactionInfoDTO
     */
    transaction: EmbeddedAccountKeyLinkTransactionDTO | EmbeddedNodeKeyLinkTransactionDTO | EmbeddedVrfKeyLinkTransactionDTO | EmbeddedVotingKeyLinkTransactionDTO | EmbeddedHashLockTransactionDTO | EmbeddedSecretLockTransactionDTO | EmbeddedSecretProofTransactionDTO | EmbeddedAccountMetadataTransactionDTO | EmbeddedMosaicMetadataTransactionDTO | EmbeddedNamespaceMetadataTransactionDTO | EmbeddedMosaicDefinitionTransactionDTO | EmbeddedMosaicSupplyChangeTransactionDTO | EmbeddedNamespaceRegistrationTransactionDTO | EmbeddedAddressAliasTransactionDTO | EmbeddedMosaicAliasTransactionDTO | EmbeddedMultisigAccountModificationTransactionDTO | EmbeddedAccountAddressRestrictionTransactionDTO | EmbeddedAccountMosaicRestrictionTransactionDTO | EmbeddedAccountOperationRestrictionTransactionDTO | EmbeddedMosaicGlobalRestrictionTransactionDTO | EmbeddedMosaicAddressRestrictionTransactionDTO | EmbeddedTransferTransactionDTO;
}

export function EmbeddedTransactionInfoDTOFromJSON(json: any): EmbeddedTransactionInfoDTO {
    return EmbeddedTransactionInfoDTOFromJSONTyped(json, false);
}

export function EmbeddedTransactionInfoDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): EmbeddedTransactionInfoDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'id': json['id'],
        'meta': json['meta'],
        'transaction': json['transaction'],
    };
}

export function EmbeddedTransactionInfoDTOToJSON(value?: EmbeddedTransactionInfoDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'id': value.id,
        'meta': value.meta,
        'transaction': value.transaction,
    };
}
