// symbol-openapi-typescript-fetch-client/src/models/EmbeddedSecretLockTransactionDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    EmbeddedTransactionDTO,
    EmbeddedTransactionDTOFromJSON,
    EmbeddedTransactionDTOFromJSONTyped,
    EmbeddedTransactionDTOToJSON,
    LockHashAlgorithmEnum,
    LockHashAlgorithmEnumFromJSON,
    LockHashAlgorithmEnumFromJSONTyped,
    LockHashAlgorithmEnumToJSON,
    NetworkTypeEnum,
    NetworkTypeEnumFromJSON,
    NetworkTypeEnumFromJSONTyped,
    NetworkTypeEnumToJSON,
    SecretLockTransactionBodyDTO,
    SecretLockTransactionBodyDTOFromJSON,
    SecretLockTransactionBodyDTOFromJSONTyped,
    SecretLockTransactionBodyDTOToJSON,
} from './';

/**
 * 
 * @export
 * @interface EmbeddedSecretLockTransactionDTO
 */
export interface EmbeddedSecretLockTransactionDTO {
    /**
     * Public key.
     * @type {string}
     * @memberof EmbeddedSecretLockTransactionDTO
     */
    signerPublicKey: string;
    /**
     * Entity version.
     * @type {number}
     * @memberof EmbeddedSecretLockTransactionDTO
     */
    version: number;
    /**
     * 
     * @type {NetworkTypeEnum}
     * @memberof EmbeddedSecretLockTransactionDTO
     */
    network: NetworkTypeEnum;
    /**
     * 
     * @type {number}
     * @memberof EmbeddedSecretLockTransactionDTO
     */
    type: number;
    /**
     * Address expressed in hexadecimal base. If the bit 0 of byte 0 is not set (like in 0x90), then it is a
     * regular address. Else (e.g. 0x91) it represents a namespace id which starts at byte 1.
     * @type {string}
     * @memberof EmbeddedSecretLockTransactionDTO
     */
    recipientAddress: string;
    /**
     * 
     * @type {string}
     * @memberof EmbeddedSecretLockTransactionDTO
     */
    secret: string;
    /**
     * Mosaic identifier. If the most significant bit of byte 0 is set, a namespaceId (alias)
     * is used instead of the real mosaic identifier.
     * @type {string}
     * @memberof EmbeddedSecretLockTransactionDTO
     */
    mosaicId: string;
    /**
     * Absolute amount. An amount of 123456789 (absolute) for a mosaic with divisibility 6 means 123.456789 (relative).
     * @type {string}
     * @memberof EmbeddedSecretLockTransactionDTO
     */
    amount: string;
    /**
     * Duration expressed in number of blocks.
     * @type {string}
     * @memberof EmbeddedSecretLockTransactionDTO
     */
    duration: string;
    /**
     * 
     * @type {LockHashAlgorithmEnum}
     * @memberof EmbeddedSecretLockTransactionDTO
     */
    hashAlgorithm: LockHashAlgorithmEnum;
}

export function EmbeddedSecretLockTransactionDTOFromJSON(json: any): EmbeddedSecretLockTransactionDTO {
    return EmbeddedSecretLockTransactionDTOFromJSONTyped(json, false);
}

export function EmbeddedSecretLockTransactionDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): EmbeddedSecretLockTransactionDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'signerPublicKey': json['signerPublicKey'],
        'version': json['version'],
        'network': json['network'],
        'type': json['type'],
        'recipientAddress': json['recipientAddress'],
        'secret': json['secret'],
        'mosaicId': json['mosaicId'],
        'amount': json['amount'],
        'duration': json['duration'],
        'hashAlgorithm': json['hashAlgorithm'],
    };
}

export function EmbeddedSecretLockTransactionDTOToJSON(value?: EmbeddedSecretLockTransactionDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'signerPublicKey': value.signerPublicKey,
        'version': value.version,
        'network': value.network,
        'type': value.type,
        'recipientAddress': value.recipientAddress,
        'secret': value.secret,
        'mosaicId': value.mosaicId,
        'amount': value.amount,
        'duration': value.duration,
        'hashAlgorithm': value.hashAlgorithm,
    };
}
