// symbol-openapi-typescript-fetch-client/src/models/EmbeddedMosaicAddressRestrictionTransactionDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    EmbeddedTransactionDTO,
    EmbeddedTransactionDTOFromJSON,
    EmbeddedTransactionDTOFromJSONTyped,
    EmbeddedTransactionDTOToJSON,
    MosaicAddressRestrictionTransactionBodyDTO,
    MosaicAddressRestrictionTransactionBodyDTOFromJSON,
    MosaicAddressRestrictionTransactionBodyDTOFromJSONTyped,
    MosaicAddressRestrictionTransactionBodyDTOToJSON,
    NetworkTypeEnum,
    NetworkTypeEnumFromJSON,
    NetworkTypeEnumFromJSONTyped,
    NetworkTypeEnumToJSON,
} from './';

/**
 * 
 * @export
 * @interface EmbeddedMosaicAddressRestrictionTransactionDTO
 */
export interface EmbeddedMosaicAddressRestrictionTransactionDTO {
    /**
     * Public key.
     * @type {string}
     * @memberof EmbeddedMosaicAddressRestrictionTransactionDTO
     */
    signerPublicKey: string;
    /**
     * Entity version.
     * @type {number}
     * @memberof EmbeddedMosaicAddressRestrictionTransactionDTO
     */
    version: number;
    /**
     * 
     * @type {NetworkTypeEnum}
     * @memberof EmbeddedMosaicAddressRestrictionTransactionDTO
     */
    network: NetworkTypeEnum;
    /**
     * 
     * @type {number}
     * @memberof EmbeddedMosaicAddressRestrictionTransactionDTO
     */
    type: number;
    /**
     * Mosaic identifier. If the most significant bit of byte 0 is set, a namespaceId (alias)
     * is used instead of the real mosaic identifier.
     * @type {string}
     * @memberof EmbeddedMosaicAddressRestrictionTransactionDTO
     */
    mosaicId: string;
    /**
     * Restriction key.
     * @type {string}
     * @memberof EmbeddedMosaicAddressRestrictionTransactionDTO
     */
    restrictionKey: string;
    /**
     * Restriction value.
     * @type {string}
     * @memberof EmbeddedMosaicAddressRestrictionTransactionDTO
     */
    previousRestrictionValue: string;
    /**
     * Restriction value.
     * @type {string}
     * @memberof EmbeddedMosaicAddressRestrictionTransactionDTO
     */
    newRestrictionValue: string;
    /**
     * Address expressed in hexadecimal base. If the bit 0 of byte 0 is not set (like in 0x90), then it is a
     * regular address. Else (e.g. 0x91) it represents a namespace id which starts at byte 1.
     * @type {string}
     * @memberof EmbeddedMosaicAddressRestrictionTransactionDTO
     */
    targetAddress: string;
}

export function EmbeddedMosaicAddressRestrictionTransactionDTOFromJSON(json: any): EmbeddedMosaicAddressRestrictionTransactionDTO {
    return EmbeddedMosaicAddressRestrictionTransactionDTOFromJSONTyped(json, false);
}

export function EmbeddedMosaicAddressRestrictionTransactionDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): EmbeddedMosaicAddressRestrictionTransactionDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'signerPublicKey': json['signerPublicKey'],
        'version': json['version'],
        'network': json['network'],
        'type': json['type'],
        'mosaicId': json['mosaicId'],
        'restrictionKey': json['restrictionKey'],
        'previousRestrictionValue': json['previousRestrictionValue'],
        'newRestrictionValue': json['newRestrictionValue'],
        'targetAddress': json['targetAddress'],
    };
}

export function EmbeddedMosaicAddressRestrictionTransactionDTOToJSON(value?: EmbeddedMosaicAddressRestrictionTransactionDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'signerPublicKey': value.signerPublicKey,
        'version': value.version,
        'network': value.network,
        'type': value.type,
        'mosaicId': value.mosaicId,
        'restrictionKey': value.restrictionKey,
        'previousRestrictionValue': value.previousRestrictionValue,
        'newRestrictionValue': value.newRestrictionValue,
        'targetAddress': value.targetAddress,
    };
}
