// symbol-openapi-typescript-fetch-client/src/models/FinalizedBlockDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
/**
 * 
 * @export
 * @interface FinalizedBlockDTO
 */
export interface FinalizedBlockDTO {
    /**
     * Finalization Epoch
     * @type {number}
     * @memberof FinalizedBlockDTO
     */
    finalizationEpoch: number;
    /**
     * Finalization point
     * @type {number}
     * @memberof FinalizedBlockDTO
     */
    finalizationPoint: number;
    /**
     * Height of the blockchain.
     * @type {string}
     * @memberof FinalizedBlockDTO
     */
    height: string;
    /**
     * 
     * @type {string}
     * @memberof FinalizedBlockDTO
     */
    hash: string;
}

export function FinalizedBlockDTOFromJSON(json: any): FinalizedBlockDTO {
    return FinalizedBlockDTOFromJSONTyped(json, false);
}

export function FinalizedBlockDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): FinalizedBlockDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'finalizationEpoch': json['finalizationEpoch'],
        'finalizationPoint': json['finalizationPoint'],
        'height': json['height'],
        'hash': json['hash'],
    };
}

export function FinalizedBlockDTOToJSON(value?: FinalizedBlockDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'finalizationEpoch': value.finalizationEpoch,
        'finalizationPoint': value.finalizationPoint,
        'height': value.height,
        'hash': value.hash,
    };
}
