// symbol-openapi-typescript-fetch-client/src/models/HashLockNetworkPropertiesDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
/**
 * 
 * @export
 * @interface HashLockNetworkPropertiesDTO
 */
export interface HashLockNetworkPropertiesDTO {
    /**
     * Amount that has to be locked per aggregate in partial cache.
     * @type {string}
     * @memberof HashLockNetworkPropertiesDTO
     */
    lockedFundsPerAggregate?: string;
    /**
     * Maximum number of blocks for which a hash lock can exist.
     * @type {string}
     * @memberof HashLockNetworkPropertiesDTO
     */
    maxHashLockDuration?: string;
}

export function HashLockNetworkPropertiesDTOFromJSON(json: any): HashLockNetworkPropertiesDTO {
    return HashLockNetworkPropertiesDTOFromJSONTyped(json, false);
}

export function HashLockNetworkPropertiesDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): HashLockNetworkPropertiesDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'lockedFundsPerAggregate': !exists(json, 'lockedFundsPerAggregate') ? undefined : json['lockedFundsPerAggregate'],
        'maxHashLockDuration': !exists(json, 'maxHashLockDuration') ? undefined : json['maxHashLockDuration'],
    };
}

export function HashLockNetworkPropertiesDTOToJSON(value?: HashLockNetworkPropertiesDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'lockedFundsPerAggregate': value.lockedFundsPerAggregate,
        'maxHashLockDuration': value.maxHashLockDuration,
    };
}
