// symbol-openapi-typescript-fetch-client/src/models/MetadataInfoDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    MetadataEntryDTO,
    MetadataEntryDTOFromJSON,
    MetadataEntryDTOFromJSONTyped,
    MetadataEntryDTOToJSON,
} from './';

/**
 * 
 * @export
 * @interface MetadataInfoDTO
 */
export interface MetadataInfoDTO {
    /**
     * 
     * @type {string}
     * @memberof MetadataInfoDTO
     */
    id: string;
    /**
     * 
     * @type {MetadataEntryDTO}
     * @memberof MetadataInfoDTO
     */
    metadataEntry: MetadataEntryDTO;
}

export function MetadataInfoDTOFromJSON(json: any): MetadataInfoDTO {
    return MetadataInfoDTOFromJSONTyped(json, false);
}

export function MetadataInfoDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): MetadataInfoDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'id': json['id'],
        'metadataEntry': json['metadataEntry'],
    };
}

export function MetadataInfoDTOToJSON(value?: MetadataInfoDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'id': value.id,
        'metadataEntry': value.metadataEntry,
    };
}
