// symbol-openapi-typescript-fetch-client/src/models/MosaicRestrictionsPage.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    MosaicAddressRestrictionDTO,
    MosaicAddressRestrictionDTOFromJSON,
    MosaicAddressRestrictionDTOFromJSONTyped,
    MosaicAddressRestrictionDTOToJSON,
    MosaicGlobalRestrictionDTO,
    MosaicGlobalRestrictionDTOFromJSON,
    MosaicGlobalRestrictionDTOFromJSONTyped,
    MosaicGlobalRestrictionDTOToJSON,
    Pagination,
    PaginationFromJSON,
    PaginationFromJSONTyped,
    PaginationToJSON,
} from './';

/**
 * 
 * @export
 * @interface MosaicRestrictionsPage
 */
export interface MosaicRestrictionsPage {
    /**
     * Array of mosaic restrictions.
     * @type {Array<MosaicAddressRestrictionDTO | MosaicGlobalRestrictionDTO>}
     * @memberof MosaicRestrictionsPage
     */
    data: Array<MosaicAddressRestrictionDTO | MosaicGlobalRestrictionDTO>;
    /**
     * 
     * @type {Pagination}
     * @memberof MosaicRestrictionsPage
     */
    pagination: Pagination;
}

export function MosaicRestrictionsPageFromJSON(json: any): MosaicRestrictionsPage {
    return MosaicRestrictionsPageFromJSONTyped(json, false);
}

export function MosaicRestrictionsPageFromJSONTyped(json: any, ignoreDiscriminator: boolean): MosaicRestrictionsPage {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'data': ((json['data'] as Array<any>)),
        'pagination': json['pagination'],
    };
}

export function MosaicRestrictionsPageToJSON(value?: MosaicRestrictionsPage | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'data': ((value.data as Array<any>)),
        'pagination': value.pagination,
    };
}
