// symbol-openapi-typescript-fetch-client/src/models/MosaicDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
/**
 * 
 * @export
 * @interface MosaicDTO
 */
export interface MosaicDTO {
    /**
     * Mosaic identifier.
     * @type {string}
     * @memberof MosaicDTO
     */
    id: string;
    /**
     * Absolute amount. An amount of 123456789 (absolute) for a mosaic with divisibility 6 means 123.456789 (relative).
     * @type {string}
     * @memberof MosaicDTO
     */
    supply: string;
    /**
     * Height of the blockchain.
     * @type {string}
     * @memberof MosaicDTO
     */
    startHeight: string;
    /**
     * Address expressed in hexadecimal base.
     * @type {string}
     * @memberof MosaicDTO
     */
    ownerAddress: string;
    /**
     * A number that allows uint 32 values.
     * @type {number}
     * @memberof MosaicDTO
     */
    revision: number;
    /**
     * - 0x00 (none) - No flags present.
     * - 0x01 (supplyMutable) - Mosaic supports supply changes even when mosaic owner owns partial supply.
     * - 0x02 (transferable) - Mosaic supports transfers between arbitrary accounts. When not set, mosaic can only be transferred to and from mosaic owner.
     * - 0x04 (restrictable) - Mosaic supports custom restrictions configured by mosaic owner.
     * @type {number}
     * @memberof MosaicDTO
     */
    flags: number;
    /**
     * Determines up to what decimal place the mosaic can be divided.
     * Divisibility of 3 means that a mosaic can be divided into smallest parts of 0.001 mosaics.
     * The divisibility must be in the range of 0 and 6.
     * @type {number}
     * @memberof MosaicDTO
     */
    divisibility: number;
    /**
     * Duration expressed in number of blocks.
     * @type {string}
     * @memberof MosaicDTO
     */
    duration: string;
}

export function MosaicDTOFromJSON(json: any): MosaicDTO {
    return MosaicDTOFromJSONTyped(json, false);
}

export function MosaicDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): MosaicDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'id': json['id'],
        'supply': json['supply'],
        'startHeight': json['startHeight'],
        'ownerAddress': json['ownerAddress'],
        'revision': json['revision'],
        'flags': json['flags'],
        'divisibility': json['divisibility'],
        'duration': json['duration'],
    };
}

export function MosaicDTOToJSON(value?: MosaicDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'id': value.id,
        'supply': value.supply,
        'startHeight': value.startHeight,
        'ownerAddress': value.ownerAddress,
        'revision': value.revision,
        'flags': value.flags,
        'divisibility': value.divisibility,
        'duration': value.duration,
    };
}
