// symbol-openapi-typescript-fetch-client/src/models/MosaicNetworkPropertiesDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
/**
 * 
 * @export
 * @interface MosaicNetworkPropertiesDTO
 */
export interface MosaicNetworkPropertiesDTO {
    /**
     * Maximum number of mosaics that an account can own.
     * @type {string}
     * @memberof MosaicNetworkPropertiesDTO
     */
    maxMosaicsPerAccount?: string;
    /**
     * Maximum mosaic duration.
     * @type {string}
     * @memberof MosaicNetworkPropertiesDTO
     */
    maxMosaicDuration?: string;
    /**
     * Maximum mosaic divisibility.
     * @type {string}
     * @memberof MosaicNetworkPropertiesDTO
     */
    maxMosaicDivisibility?: string;
    /**
     * Address encoded using a 32-character set.
     * @type {string}
     * @memberof MosaicNetworkPropertiesDTO
     */
    mosaicRentalFeeSinkAddress?: string;
    /**
     * Mosaic rental fee.
     * @type {string}
     * @memberof MosaicNetworkPropertiesDTO
     */
    mosaicRentalFee?: string;
}

export function MosaicNetworkPropertiesDTOFromJSON(json: any): MosaicNetworkPropertiesDTO {
    return MosaicNetworkPropertiesDTOFromJSONTyped(json, false);
}

export function MosaicNetworkPropertiesDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): MosaicNetworkPropertiesDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'maxMosaicsPerAccount': !exists(json, 'maxMosaicsPerAccount') ? undefined : json['maxMosaicsPerAccount'],
        'maxMosaicDuration': !exists(json, 'maxMosaicDuration') ? undefined : json['maxMosaicDuration'],
        'maxMosaicDivisibility': !exists(json, 'maxMosaicDivisibility') ? undefined : json['maxMosaicDivisibility'],
        'mosaicRentalFeeSinkAddress': !exists(json, 'mosaicRentalFeeSinkAddress') ? undefined : json['mosaicRentalFeeSinkAddress'],
        'mosaicRentalFee': !exists(json, 'mosaicRentalFee') ? undefined : json['mosaicRentalFee'],
    };
}

export function MosaicNetworkPropertiesDTOToJSON(value?: MosaicNetworkPropertiesDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'maxMosaicsPerAccount': value.maxMosaicsPerAccount,
        'maxMosaicDuration': value.maxMosaicDuration,
        'maxMosaicDivisibility': value.maxMosaicDivisibility,
        'mosaicRentalFeeSinkAddress': value.mosaicRentalFeeSinkAddress,
        'mosaicRentalFee': value.mosaicRentalFee,
    };
}
