// symbol-openapi-typescript-fetch-client/src/models/NamespaceRegistrationTransactionDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    NamespaceRegistrationTransactionBodyDTO,
    NamespaceRegistrationTransactionBodyDTOFromJSON,
    NamespaceRegistrationTransactionBodyDTOFromJSONTyped,
    NamespaceRegistrationTransactionBodyDTOToJSON,
    NamespaceRegistrationTypeEnum,
    NamespaceRegistrationTypeEnumFromJSON,
    NamespaceRegistrationTypeEnumFromJSONTyped,
    NamespaceRegistrationTypeEnumToJSON,
    NetworkTypeEnum,
    NetworkTypeEnumFromJSON,
    NetworkTypeEnumFromJSONTyped,
    NetworkTypeEnumToJSON,
    TransactionDTO,
    TransactionDTOFromJSON,
    TransactionDTOFromJSONTyped,
    TransactionDTOToJSON,
} from './';

/**
 * Transaction to create or renew a namespace.
 * @export
 * @interface NamespaceRegistrationTransactionDTO
 */
export interface NamespaceRegistrationTransactionDTO {
    /**
     * A number that allows uint 32 values.
     * @type {number}
     * @memberof NamespaceRegistrationTransactionDTO
     */
    size: number;
    /**
     * Entity's signature generated by the signer.
     * @type {string}
     * @memberof NamespaceRegistrationTransactionDTO
     */
    signature: string;
    /**
     * Public key.
     * @type {string}
     * @memberof NamespaceRegistrationTransactionDTO
     */
    signerPublicKey: string;
    /**
     * Entity version.
     * @type {number}
     * @memberof NamespaceRegistrationTransactionDTO
     */
    version: number;
    /**
     * 
     * @type {NetworkTypeEnum}
     * @memberof NamespaceRegistrationTransactionDTO
     */
    network: NetworkTypeEnum;
    /**
     * 
     * @type {number}
     * @memberof NamespaceRegistrationTransactionDTO
     */
    type: number;
    /**
     * Absolute amount. An amount of 123456789 (absolute) for a mosaic with divisibility 6 means 123.456789 (relative).
     * @type {string}
     * @memberof NamespaceRegistrationTransactionDTO
     */
    maxFee: string;
    /**
     * Duration expressed in number of blocks.
     * @type {string}
     * @memberof NamespaceRegistrationTransactionDTO
     */
    deadline: string;
    /**
     * Duration expressed in number of blocks.
     * @type {string}
     * @memberof NamespaceRegistrationTransactionDTO
     */
    duration: string;
    /**
     * Namespace identifier.
     * @type {string}
     * @memberof NamespaceRegistrationTransactionDTO
     */
    parentId: string;
    /**
     * Namespace identifier.
     * @type {string}
     * @memberof NamespaceRegistrationTransactionDTO
     */
    id: string;
    /**
     * 
     * @type {NamespaceRegistrationTypeEnum}
     * @memberof NamespaceRegistrationTransactionDTO
     */
    registrationType: NamespaceRegistrationTypeEnum;
    /**
     * Namespace name.
     * @type {string}
     * @memberof NamespaceRegistrationTransactionDTO
     */
    name: string;
}

export function NamespaceRegistrationTransactionDTOFromJSON(json: any): NamespaceRegistrationTransactionDTO {
    return NamespaceRegistrationTransactionDTOFromJSONTyped(json, false);
}

export function NamespaceRegistrationTransactionDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): NamespaceRegistrationTransactionDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'size': json['size'],
        'signature': json['signature'],
        'signerPublicKey': json['signerPublicKey'],
        'version': json['version'],
        'network': json['network'],
        'type': json['type'],
        'maxFee': json['maxFee'],
        'deadline': json['deadline'],
        'duration': json['duration'],
        'parentId': json['parentId'],
        'id': json['id'],
        'registrationType': json['registrationType'],
        'name': json['name'],
    };
}

export function NamespaceRegistrationTransactionDTOToJSON(value?: NamespaceRegistrationTransactionDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'size': value.size,
        'signature': value.signature,
        'signerPublicKey': value.signerPublicKey,
        'version': value.version,
        'network': value.network,
        'type': value.type,
        'maxFee': value.maxFee,
        'deadline': value.deadline,
        'duration': value.duration,
        'parentId': value.parentId,
        'id': value.id,
        'registrationType': value.registrationType,
        'name': value.name,
    };
}
