// symbol-openapi-typescript-fetch-client/src/models/NodeKeyLinkTransactionBodyDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
import {
    LinkActionEnum,
    LinkActionEnumFromJSON,
    LinkActionEnumFromJSONTyped,
    LinkActionEnumToJSON,
} from './';

/**
 * 
 * @export
 * @interface NodeKeyLinkTransactionBodyDTO
 */
export interface NodeKeyLinkTransactionBodyDTO {
    /**
     * Public key.
     * @type {string}
     * @memberof NodeKeyLinkTransactionBodyDTO
     */
    linkedPublicKey: string;
    /**
     * 
     * @type {LinkActionEnum}
     * @memberof NodeKeyLinkTransactionBodyDTO
     */
    linkAction: LinkActionEnum;
}

export function NodeKeyLinkTransactionBodyDTOFromJSON(json: any): NodeKeyLinkTransactionBodyDTO {
    return NodeKeyLinkTransactionBodyDTOFromJSONTyped(json, false);
}

export function NodeKeyLinkTransactionBodyDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): NodeKeyLinkTransactionBodyDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'linkedPublicKey': json['linkedPublicKey'],
        'linkAction': json['linkAction'],
    };
}

export function NodeKeyLinkTransactionBodyDTOToJSON(value?: NodeKeyLinkTransactionBodyDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'linkedPublicKey': value.linkedPublicKey,
        'linkAction': value.linkAction,
    };
}
