// symbol-openapi-typescript-fetch-client/src/models/RolesTypeEnum.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

/**
 * Role of the node:
 * * 1 - Peer node.
 * * 2 - Api node.
 * * 3 - Peer Api node.
 * * 4 - Voting node.
 * * 5 - Peer Voting node.
 * * 6 - Api Voting node.
 * * 7 - Peer Api Voting node.
 * @export
 * @enum {string}
 */
export enum RolesTypeEnum {
    NUMBER_1 = 1,
    NUMBER_2 = 2,
    NUMBER_3 = 3,
    NUMBER_4 = 4,
    NUMBER_5 = 5,
    NUMBER_6 = 6,
    NUMBER_7 = 7
}

export function RolesTypeEnumFromJSON(json: any): RolesTypeEnum {
    return RolesTypeEnumFromJSONTyped(json, false);
}

export function RolesTypeEnumFromJSONTyped(json: any, ignoreDiscriminator: boolean): RolesTypeEnum {
    return json as RolesTypeEnum;
}

export function RolesTypeEnumToJSON(value?: RolesTypeEnum | null): any {
    return value as any;
}
