// symbol-openapi-typescript-fetch-client/src/models/StorageInfoDTO.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 2.1.0
 *
 * The version of the OpenAPI document: 0.10.0
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */

import { exists, mapValues } from '../runtime';
/**
 * 
 * @export
 * @interface StorageInfoDTO
 */
export interface StorageInfoDTO {
    /**
     * Number of blocks stored.
     * @type {number}
     * @memberof StorageInfoDTO
     */
    numBlocks: number;
    /**
     * Number of transactions stored.
     * @type {number}
     * @memberof StorageInfoDTO
     */
    numTransactions: number;
    /**
     * Number of accounts created.
     * @type {number}
     * @memberof StorageInfoDTO
     */
    numAccounts: number;
}

export function StorageInfoDTOFromJSON(json: any): StorageInfoDTO {
    return StorageInfoDTOFromJSONTyped(json, false);
}

export function StorageInfoDTOFromJSONTyped(json: any, ignoreDiscriminator: boolean): StorageInfoDTO {
    if ((json === undefined) || (json === null)) {
        return json;
    }
    return {
        
        'numBlocks': json['numBlocks'],
        'numTransactions': json['numTransactions'],
        'numAccounts': json['numAccounts'],
    };
}

export function StorageInfoDTOToJSON(value?: StorageInfoDTO | null): any {
    if (value === undefined) {
        return undefined;
    }
    if (value === null) {
        return null;
    }
    return {
        
        'numBlocks': value.numBlocks,
        'numTransactions': value.numTransactions,
        'numAccounts': value.numAccounts,
    };
}
