// catbuffer-typescript/src/MosaicResolutionStatementBuilder.ts
// tslint:disable: jsdoc-format
/**
*** Copyright (c) 2016-present,
*** Jaguar0625, gimre, BloodyRookie, Tech Bureau, Corp. All rights reserved.
***
*** This file is part of Catapult.
***
*** Catapult is free software: you can redistribute it and/or modify
*** it under the terms of the GNU Lesser General Public License as published by
*** the Free Software Foundation, either version 3 of the License, or
*** (at your option) any later version.
***
*** Catapult is distributed in the hope that it will be useful,
*** but WITHOUT ANY WARRANTY; without even the implied warranty of
*** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
*** GNU Lesser General Public License for more details.
***
*** You should have received a copy of the GNU Lesser General Public License
*** along with Catapult. If not, see <http://www.gnu.org/licenses/>.
**/

import { GeneratorUtils } from './GeneratorUtils';
import { MosaicResolutionEntryBuilder } from './MosaicResolutionEntryBuilder';
import { ReceiptBuilder } from './ReceiptBuilder';
import { ReceiptTypeDto } from './ReceiptTypeDto';
import { UnresolvedMosaicIdDto } from './UnresolvedMosaicIdDto';

/** Binary layout for a mosaic resolution statement. */
export class MosaicResolutionStatementBuilder extends ReceiptBuilder {
    /** Unresolved mosaic. */
    unresolved: UnresolvedMosaicIdDto;
    /** Resolution entries. */
    resolutionEntries: MosaicResolutionEntryBuilder[];

    /**
     * Constructor.
     *
     * @param version Receipt version.
     * @param type Receipt type.
     * @param unresolved Unresolved mosaic.
     * @param resolutionEntries Resolution entries.
     */
    // tslint:disable-next-line: max-line-length
    public constructor(version: number,  type: ReceiptTypeDto,  unresolved: UnresolvedMosaicIdDto,  resolutionEntries: MosaicResolutionEntryBuilder[]) {
        super(version, type);
        this.unresolved = unresolved;
        this.resolutionEntries = resolutionEntries;
    }

    /**
     * Creates an instance of MosaicResolutionStatementBuilder from binary payload.
     *
     * @param payload Byte payload to use to serialize the object.
     * @return Instance of MosaicResolutionStatementBuilder.
     */
    public static loadFromBinary(payload: Uint8Array): MosaicResolutionStatementBuilder {
        const byteArray = Array.from(payload);
        const superObject = ReceiptBuilder.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, superObject.getSize());
        const unresolved = UnresolvedMosaicIdDto.loadFromBinary(Uint8Array.from(byteArray));
        byteArray.splice(0, unresolved.getSize());
        let resolutionEntriesByteSize = byteArray.length;
        const resolutionEntries: MosaicResolutionEntryBuilder[] = [];
        while (resolutionEntriesByteSize > 0) {
            const item = MosaicResolutionEntryBuilder.loadFromBinary(Uint8Array.from(byteArray));
            resolutionEntries.push(item);
            const itemSize = item.getSize();
            resolutionEntriesByteSize -= itemSize;
            byteArray.splice(0, itemSize);
        }
        return new MosaicResolutionStatementBuilder(superObject.version, superObject.type, unresolved, resolutionEntries);
    }

    /**
     * Gets unresolved mosaic.
     *
     * @return Unresolved mosaic.
     */
    public getUnresolved(): UnresolvedMosaicIdDto {
        return this.unresolved;
    }

    /**
     * Gets resolution entries.
     *
     * @return Resolution entries.
     */
    public getResolutionEntries(): MosaicResolutionEntryBuilder[] {
        return this.resolutionEntries;
    }

    /**
     * Gets the size of the object.
     *
     * @return Size in bytes.
     */
    public getSize(): number {
        let size: number = super.getSize();
        size += this.unresolved.getSize();
        this.resolutionEntries.forEach((o) => size += o.getSize());
        return size;
    }

    /**
     * Serializes an object to bytes.
     *
     * @return Serialized bytes.
     */
    public serialize(): Uint8Array {
        let newArray = Uint8Array.from([]);
        const superBytes = super.serialize();
        newArray = GeneratorUtils.concatTypedArrays(newArray, superBytes);
        const unresolvedBytes = this.unresolved.serialize();
        newArray = GeneratorUtils.concatTypedArrays(newArray, unresolvedBytes);
        this.resolutionEntries.forEach((item) => {
            const resolutionEntriesBytes = item.serialize();
            newArray = GeneratorUtils.concatTypedArrays(newArray, resolutionEntriesBytes);
        });
        return newArray;
    }
}
