// catbuffer-typescript/src/ProofScalarDto.ts
// tslint:disable: jsdoc-format
/**
*** Copyright (c) 2016-present,
*** Jaguar0625, gimre, BloodyRookie, Tech Bureau, Corp. All rights reserved.
***
*** This file is part of Catapult.
***
*** Catapult is free software: you can redistribute it and/or modify
*** it under the terms of the GNU Lesser General Public License as published by
*** the Free Software Foundation, either version 3 of the License, or
*** (at your option) any later version.
***
*** Catapult is distributed in the hope that it will be useful,
*** but WITHOUT ANY WARRANTY; without even the implied warranty of
*** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
*** GNU Lesser General Public License for more details.
***
*** You should have received a copy of the GNU Lesser General Public License
*** along with Catapult. If not, see <http://www.gnu.org/licenses/>.
**/

import { GeneratorUtils } from './GeneratorUtils';

/** Proof scalar. */
export class ProofScalarDto {
    /** Proof scalar. */
    proofScalar: Uint8Array;

    /**
     * Constructor.
     *
     * @param proofScalar Proof scalar.
     */
     constructor(proofScalar: Uint8Array) {
        this.proofScalar = proofScalar;
    }

    /**
     * Creates an instance of ProofScalarDto from binary payload.
     *
     * @param payload Byte payload to use to serialize the object.
     * @return Instance of ProofScalarDto.
     */
    public static loadFromBinary(payload: Uint8Array): ProofScalarDto {
        const byteArray = Array.from(payload);
        const proofScalar = GeneratorUtils.getBytes(Uint8Array.from(byteArray), 32);
        byteArray.splice(0, 32);
        return new ProofScalarDto(proofScalar);
    }

    /**
     * Gets Proof scalar.
     *
     * @return Proof scalar.
     */
    public getProofScalar(): Uint8Array {
        return this.proofScalar;
    }

    /**
     * Gets the size of the object.
     *
     * @return Size in bytes.
     */
    public getSize(): number {
        return 32;
    }

    /**
     * Serializes an object to bytes.
     *
     * @return Serialized bytes.
     */
    public serialize(): Uint8Array {
        let newArray = Uint8Array.from([]);
        newArray = GeneratorUtils.concatTypedArrays(newArray, this.proofScalar);
        return newArray;
    }
}
