// catbuffer-typescript/src/ReceiptSourceBuilder.ts
// tslint:disable: jsdoc-format
/**
*** Copyright (c) 2016-present,
*** Jaguar0625, gimre, BloodyRookie, Tech Bureau, Corp. All rights reserved.
***
*** This file is part of Catapult.
***
*** Catapult is free software: you can redistribute it and/or modify
*** it under the terms of the GNU Lesser General Public License as published by
*** the Free Software Foundation, either version 3 of the License, or
*** (at your option) any later version.
***
*** Catapult is distributed in the hope that it will be useful,
*** but WITHOUT ANY WARRANTY; without even the implied warranty of
*** MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
*** GNU Lesser General Public License for more details.
***
*** You should have received a copy of the GNU Lesser General Public License
*** along with Catapult. If not, see <http://www.gnu.org/licenses/>.
**/

import { GeneratorUtils } from './GeneratorUtils';

/** Binary layout for receipt source. */
export class ReceiptSourceBuilder {
    /** Transaction primary source (e.g. index within block). */
    primaryId: number;
    /** Transaction secondary source (e.g. index within aggregate). */
    secondaryId: number;

    /**
     * Constructor.
     *
     * @param primaryId Transaction primary source (e.g. index within block).
     * @param secondaryId Transaction secondary source (e.g. index within aggregate).
     */
    public constructor(primaryId: number,  secondaryId: number) {
        this.primaryId = primaryId;
        this.secondaryId = secondaryId;
    }

    /**
     * Creates an instance of ReceiptSourceBuilder from binary payload.
     *
     * @param payload Byte payload to use to serialize the object.
     * @return Instance of ReceiptSourceBuilder.
     */
    public static loadFromBinary(payload: Uint8Array): ReceiptSourceBuilder {
        const byteArray = Array.from(payload);
        const primaryId = GeneratorUtils.bufferToUint(GeneratorUtils.getBytes(Uint8Array.from(byteArray), 4));
        byteArray.splice(0, 4);
        const secondaryId = GeneratorUtils.bufferToUint(GeneratorUtils.getBytes(Uint8Array.from(byteArray), 4));
        byteArray.splice(0, 4);
        return new ReceiptSourceBuilder(primaryId, secondaryId);
    }

    /**
     * Gets transaction primary source (e.g. index within block).
     *
     * @return Transaction primary source (e.g. index within block).
     */
    public getPrimaryId(): number {
        return this.primaryId;
    }

    /**
     * Gets transaction secondary source (e.g. index within aggregate).
     *
     * @return Transaction secondary source (e.g. index within aggregate).
     */
    public getSecondaryId(): number {
        return this.secondaryId;
    }

    /**
     * Gets the size of the object.
     *
     * @return Size in bytes.
     */
    public getSize(): number {
        let size = 0;
        size += 4; // primaryId
        size += 4; // secondaryId
        return size;
    }

    /**
     * Serializes an object to bytes.
     *
     * @return Serialized bytes.
     */
    public serialize(): Uint8Array {
        let newArray = Uint8Array.from([]);
        const primaryIdBytes = GeneratorUtils.uintToBuffer(this.getPrimaryId(), 4);
        newArray = GeneratorUtils.concatTypedArrays(newArray, primaryIdBytes);
        const secondaryIdBytes = GeneratorUtils.uintToBuffer(this.getSecondaryId(), 4);
        newArray = GeneratorUtils.concatTypedArrays(newArray, secondaryIdBytes);
        return newArray;
    }
}
