// symbol-openapi-typescript-fetch-client/src/apis/ChainRoutesApi.ts
/* tslint:disable */
/* eslint-disable */
/*
 * Copyright 2019 NEM
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
/**
 * Catapult REST Endpoints
 * OpenAPI Specification of catapult-rest 1.2.0
 *
 * The version of the OpenAPI document: 0.9.6
 * 
 *
 * NOTE: This class is auto generated by OpenAPI Generator (https://openapi-generator.tech).
 * https://openapi-generator.tech
 * Do not edit the class manually.
 */


import * as runtime from '../runtime';
import {
    ChainScoreDTO,
    ChainScoreDTOFromJSON,
    ChainScoreDTOToJSON,
    HeightInfoDTO,
    HeightInfoDTOFromJSON,
    HeightInfoDTOToJSON,
} from '../models';

/**
 * 
 */
export class ChainRoutesApi extends runtime.BaseAPI {

    /**
     * Returns the current height of the blockchain.
     * Get the current height of the chain
     */
    async getChainHeightRaw(): Promise<runtime.ApiResponse<HeightInfoDTO>> {
        const queryParameters: runtime.HTTPQuery = {};

        const headerParameters: runtime.HTTPHeaders = {};

        const response = await this.request({
            path: `/chain/height`,
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        });

        return new runtime.JSONApiResponse(response, (jsonValue) => HeightInfoDTOFromJSON(jsonValue));
    }

    /**
     * Returns the current height of the blockchain.
     * Get the current height of the chain
     */
    async getChainHeight(): Promise<HeightInfoDTO> {
        const response = await this.getChainHeightRaw();
        return await response.value();
    }

    /**
     * Gets the current score of the blockchain. The higher the score, the better the chain. During synchronization, nodes try to get the best blockchain in the network.  The score for a block is derived from its difficulty and the time (in seconds) that has elapsed since the last block:      block score = difficulty − time elapsed since last block 
     * Get the current score of the chain
     */
    async getChainScoreRaw(): Promise<runtime.ApiResponse<ChainScoreDTO>> {
        const queryParameters: runtime.HTTPQuery = {};

        const headerParameters: runtime.HTTPHeaders = {};

        const response = await this.request({
            path: `/chain/score`,
            method: 'GET',
            headers: headerParameters,
            query: queryParameters,
        });

        return new runtime.JSONApiResponse(response, (jsonValue) => ChainScoreDTOFromJSON(jsonValue));
    }

    /**
     * Gets the current score of the blockchain. The higher the score, the better the chain. During synchronization, nodes try to get the best blockchain in the network.  The score for a block is derived from its difficulty and the time (in seconds) that has elapsed since the last block:      block score = difficulty − time elapsed since last block 
     * Get the current score of the chain
     */
    async getChainScore(): Promise<ChainScoreDTO> {
        const response = await this.getChainScoreRaw();
        return await response.value();
    }

}
